<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Velocity review" href="03-velocity-review.html" /><link rel="prev" title="Data structures review" href="01-data-structures-review.html" />

    <meta name="generator" content="sphinx-5.2.2, furo 2022.09.15"/>
        <title>Analysis review - Single cell analysis</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=9ec31e2665bf879c1d47d93a8ec4893870ee1e45" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">Single cell analysis</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">Single cell analysis</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="01-data-structures-review.html">Data structures review</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Analysis review</a></li>
<li class="toctree-l1"><a class="reference internal" href="03-velocity-review.html">Velocity review</a></li>
<li class="toctree-l1"><a class="reference internal" href="plotting.html">Plotting setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/cameronraysmith/sc/edit/master/source/02-analysis-review.ipynb" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="admonition note">
  This page was generated from
  <a class="reference external" href="https://github.com/cameronraysmith/sc/blob/master/source/02-analysis-review.ipynb">source/02-analysis-review.ipynb</a>.
  <span style="white-space: nowrap;">
  <a href="https://mybinder.org/v2/gh/cameronraysmith/sc/master?filepath=source/02-analysis-review.ipynb"><img alt="Binder badge" src="https://mybinder.org/badge_logo.svg" style="vertical-align:text-bottom"></a> or
  <a href="https://colab.research.google.com/github/cameronraysmith/sc/blob/master/source/02-analysis-review.ipynb"><img alt="Colab badge" src="https://colab.research.google.com/assets/colab-badge.svg" style="vertical-align:text-bottom"></a>
  </span>
</div><div class="section" id="Analysis-review">
<h1>Analysis review<a class="headerlink" href="#Analysis-review" title="Permalink to this heading">§</a></h1>
<div class="section" id="Overview">
<h2>Overview<a class="headerlink" href="#Overview" title="Permalink to this heading">§</a></h2>
<p>This notebook follows the tutorial by <a class="reference external" href="https://github.com/mousepixels/sanbomics/blob/main/single_cell_analysis_complete_class.ipynb">mousepixels/sanbomics</a>, which has an accompanying <a class="reference external" href="https://youtu.be/uvyG9yLuNSE?t=319">screencast</a>.</p>
<p>Analysis is illustrated with single-nucleus RNA sequencing data from the following paper <span id="id1">[<a class="reference internal" href="references.html#id4" title="Johannes C Melms, Jana Biermann, Huachao Huang, Yiping Wang, Ajay Nair, Somnath Tagore, Igor Katsyv, André F Rendeiro, Amit Dipak Amin, Denis Schapiro, Chris J Frangieh, Adrienne M Luoma, Aveline Filliol, Yinshan Fang, Hiranmayi Ravichandran, Mariano G Clausi, George A Alba, Meri Rogava, Sean W Chen, Patricia Ho, Daniel T Montoro, Adam E Kornberg, Arnold S Han, Mathieu F Bakhoum, Niroshana Anandasabapathy, Mayte Suárez-Fariñas, Samuel F Bakhoum, Yaron Bram, Alain Borczuk, Xinzheng V Guo, Jay H Lefkowitch, Charles Marboe, Stephen M Lagana, Armando Del Portillo, Emily J Tsai, Emmanuel Zorn, Glen S Markowitz, Robert F Schwabe, Robert E Schwartz, Olivier Elemento, Anjali Saqi, Hanina Hibshoosh, Jianwen Que, and Benjamin Izar. A molecular single-cell lung atlas of lethal COVID-19. Nature, 595(7865):114–119, July 2021. doi:10.1038/s41586-021-03569-1.">MBH+21</a>]</span></p>
<blockquote>
<div><p>Melms JC, Biermann J, Huang H, Wang Y, Nair A, Tagore S, et al. A molecular single-cell lung atlas of lethal COVID-19. Nature. 2021;595: 114–119. <a class="reference external" href="https://doi.org/10.1038/s41586-021-03569-0">doi:10.1038/s41586-021-03569-0</a></p>
</div></blockquote>
<p>This paper examined 116,000 nuclei from the lungs of nineteen patients who underwent autopsy following death in association with COVID-19. Findings reported in the abstract of the paper include:</p>
<ol class="arabic simple">
<li><p>activated monocyte-derived macrophages and alveolar macrophages</p></li>
<li><p>impaired T cell activation</p></li>
<li><p>monocyte/macrophage-derived interleukin-1β and epithelial cell-derived interleukin-6</p></li>
<li><p>alveolar type 2 cells adopted an inflammation-associated transient progenitor cell state and failed to undergo full transition into alveolar type 1 cells</p></li>
<li><p>expansion of CTHRC1+ pathological fibroblasts</p></li>
<li><p>protein activity and ligand–receptor interactions suggest putative drug targets</p></li>
</ol>
<p>This notebook makes extensive use of <span id="id2">[<a class="reference internal" href="references.html#id6" title="F Alexander Wolf, Philipp Angerer, and Fabian J Theis. SCANPY: large-scale single-cell gene expression data analysis. Genome Biol., 19(1):15, December 2018. doi:10.1186/s13059-017-1382-0.">WAT18</a>]</span> and <span id="id3">[<a class="reference internal" href="references.html#id3" title="Romain Lopez, Jeffrey Regier, Michael B Cole, Michael I Jordan, and Nir Yosef. Deep generative modeling for single-cell transcriptomics. Nat. Methods, 15(12):1053–1058, December 2018. doi:10.1038/s41592-018-0229-2.">LRC+18</a>]</span> including updates that have been made to the underlying software packages, <a class="reference external" href="https://github.com/scverse/scanpy">scanpy</a> and <a class="reference external" href="https://github.com/scverse/scvi-tools">scvi-tools</a>, since their initial publication.</p>
</div>
<div class="section" id="Setup">
<h2>Setup<a class="headerlink" href="#Setup" title="Permalink to this heading">§</a></h2>
<div class="section" id="Import-libraries">
<h3>Import libraries<a class="headerlink" href="#Import-libraries" title="Permalink to this heading">§</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">inspect</span> <span class="kn">import</span> <span class="n">getmembers</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>
<span class="kn">from</span> <span class="nn">types</span> <span class="kn">import</span> <span class="n">FunctionType</span>

<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scvi</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Global seed set to 0
/usr/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:53: LightningDeprecationWarning: pytorch_lightning.utilities.warnings.rank_zero_deprecation has been deprecated in v1.6 and will be removed in v1.8. Use the equivalent function from the pytorch_lightning.utilities.rank_zero module instead.
  new_rank_zero_deprecation(
/usr/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:58: LightningDeprecationWarning: The `pytorch_lightning.loggers.base.rank_zero_experiment` is deprecated in v1.7 and will be removed in v1.9. Please use `pytorch_lightning.loggers.logger.rank_zero_experiment` instead.
  return new_rank_zero_deprecation(*args, **kwargs)
</pre></div></div>
</div>
</div>
<div class="section" id="Setup-plotting">
<h3>Setup plotting<a class="headerlink" href="#Setup-plotting" title="Permalink to this heading">§</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.font_manager</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1"># import matplotlib_inline</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># fonts_path = &quot;/usr/share/texmf/fonts/opentype/public/lm/&quot; #ubuntu</span>
<span class="c1"># fonts_path = &quot;~/Library/Fonts/&quot; # macos</span>
<span class="n">fonts_path</span> <span class="o">=</span> <span class="s2">&quot;/usr/share/fonts/OTF/&quot;</span>  <span class="c1"># arch</span>
<span class="c1"># user_path = &quot;$HOME/&quot; # user</span>
<span class="c1"># fonts_path = user_path + &quot;fonts/latinmodern/opentype/public/lm/&quot;  # home</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">font_manager</span><span class="o">.</span><span class="n">fontManager</span><span class="o">.</span><span class="n">addfont</span><span class="p">(</span><span class="n">fonts_path</span> <span class="o">+</span> <span class="s2">&quot;lmsans10-regular.otf&quot;</span><span class="p">)</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">font_manager</span><span class="o">.</span><span class="n">fontManager</span><span class="o">.</span><span class="n">addfont</span><span class="p">(</span><span class="n">fonts_path</span> <span class="o">+</span> <span class="s2">&quot;lmroman10-regular.otf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># https://stackoverflow.com/a/36622238/446907</span>
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_formats = [&#39;svg&#39;]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s2">&quot;default&quot;</span><span class="p">)</span>  <span class="c1"># reset default parameters</span>
<span class="c1"># https://stackoverflow.com/a/3900167/446907</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;font.size&quot;</span><span class="p">:</span> <span class="mi">16</span><span class="p">,</span>
        <span class="s2">&quot;font.family&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sans-serif&quot;</span><span class="p">],</span>
        <span class="s2">&quot;font.serif&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Latin Modern Roman&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;font.serif&quot;</span><span class="p">],</span>
        <span class="s2">&quot;font.sans-serif&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Latin Modern Sans&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;font.sans-serif&quot;</span><span class="p">],</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Utility-functions">
<h3>Utility functions<a class="headerlink" href="#Utility-functions" title="Permalink to this heading">§</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">attributes</span><span class="p">(</span><span class="n">obj</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    get object attributes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">disallowed_names</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">name</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">getmembers</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">obj</span><span class="p">))</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">FunctionType</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="n">name</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;_&quot;</span> <span class="ow">and</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">disallowed_names</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
    <span class="p">}</span>


<span class="k">def</span> <span class="nf">print_attributes</span><span class="p">(</span><span class="n">obj</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    print object attributes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pprint</span><span class="p">(</span><span class="n">attributes</span><span class="p">(</span><span class="n">obj</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Import-data">
<h2>Import data<a class="headerlink" href="#Import-data" title="Permalink to this heading">§</a></h2>
<p>Here we review how the data were downloaded, and proceed to import and inspect the data.</p>
<div class="section" id="Data-download">
<h3>Data download<a class="headerlink" href="#Data-download" title="Permalink to this heading">§</a></h3>
<p>Data with GEO accession <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171524">GSE171524</a> was downloaded using <a class="reference external" href="./data/download_geo_data.sh">./data/download_geo_data.sh</a> with parameters</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./download_geo_data.sh <span class="se">\</span>
       -a GSE132771 <span class="se">\</span>
       -f <span class="s1">&#39;ftp.*RAW.*&#39;</span> <span class="se">\</span>
       -j <span class="s1">&#39;..|.supplementary_files?|..|.url?|select(length&gt;0)&#39;</span>
</pre></div>
</div>
<p>A skeleton of this script that may work in this case is</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>!/usr/bin/env bash

<span class="c1">#-- debugging (comment to reduce stderr output)</span>
<span class="c1">#-- https://wiki.bash-hackers.org/scripting/debuggingtips</span>
<span class="nb">export</span> <span class="nv">PS4</span><span class="o">=</span><span class="s1">&#39;+(${BASH_SOURCE}:${LINENO}): ${FUNCNAME[0]:+${FUNCNAME[0]}(): }&#39;</span>
<span class="nb">set</span> -o xtrace

<span class="c1"># get metadata</span>
<span class="c1"># Melms JC, Biermann J, Huang H, Wang Y, Nair A, Tagore S, et al.</span>
<span class="c1"># A molecular single-cell lung atlas of lethal COVID-19.</span>
<span class="c1"># Nature. 2021;595: 114–119. doi:10.1038/s41586-021-03569-0</span>
<span class="c1"># GSE171524</span>
ffq -l <span class="m">1</span> -o GSE171524.json GSE171524

<span class="c1"># download raw data</span>
wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE171nnn/GSE171524/suppl/GSE171524_RAW.tar

<span class="c1"># list contents</span>
tar -tvf GSE171524_RAW.tar

<span class="c1"># untar</span>
mkdir -p GSE171524 <span class="o">&amp;&amp;</span> <span class="se">\</span>
tar -xvf GSE171524_RAW.tar -C GSE171524
</pre></div>
</div>
</div>
<div class="section" id="Data-load">
<h3>Data load<a class="headerlink" href="#Data-load" title="Permalink to this heading">§</a></h3>
<p>See the <a class="reference external" href="https://scanpy.readthedocs.io/en/latest/generated/scanpy.read_csv.html">documentation for scanpy read csv</a> which returns an <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData">AnnData object</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s2">&quot;data/GSE171524/supplementary/GSM5226574_C51ctr_raw_counts.csv.gz&quot;</span>
<span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 34546
</pre></div></div>
</div>
<p>Note the <code class="docutils literal notranslate"><span class="pre">scanpy.read_csv</span></code> function accepts gzipped files.</p>
</div>
<div class="section" id="Data-properties">
<h3>Data properties<a class="headerlink" href="#Data-properties" title="Permalink to this heading">§</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
anndata._core.anndata.AnnData
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(6099, 34546)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_attributes</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;T&#39;: AnnData object with n_obs × n_vars = 34546 × 6099,
 &#39;X&#39;: array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32),
 &#39;file&#39;: Backing file manager: no file is set.,
 &#39;filename&#39;: None,
 &#39;is_view&#39;: False,
 &#39;isbacked&#39;: False,
 &#39;isview&#39;: False,
 &#39;layers&#39;: Layers with keys: ,
 &#39;n_obs&#39;: 6099,
 &#39;n_vars&#39;: 34546,
 &#39;obs&#39;: Empty DataFrame
Columns: []
Index: [TAGGTACCATGGCCAC-1_1, ATTCACTGTAACAGGC-1_1, TAACTTCCAACCACGC-1_1, TTGGGTACACGACAAG-1_1, AGGCCACAGAGTCACG-1_1, CACTGAAGTCGAAGCA-1_1, ACTGATGTCTGCACCT-1_1, TTACCGCCACTCAGAT-1_1, TTGGTTTTCCTAGCTC-1_1, TGGGAAGTCAGTGATC-1_1, CCACGAGTCTCTTAAC-1_1, ACTTCCGCACAACGCC-1_1, GGGAAGTAGCGACCCT-1_1, TGGTAGTTCCCGTGTT-1_1, CGCATAACATGCCGGT-1_1, TCTATCACAAGGCTTT-1_1, ATCCACCAGAGGTATT-1_1, TAACGACAGATGACCG-1_1, TCTTAGTGTATGAGGC-1_1, CACTTCGCAGTACTAC-1_1, GTCAAACAGAACGTGC-1_1, GCAACCGAGGGCAGGA-1_1, CATACTTTCATCACTT-1_1, AAGAACATCGGATTAC-1_1, GGGTATTGTACGATGG-1_1, CTGTAGATCAACGTGT-1_1, GTCATTTGTATCTCGA-1_1, CCTTGTGCAGAGGGTT-1_1, AAGTTCGCAACACGTT-1_1, TCATTCACAAATCAAG-1_1, TCCATGCCAACGACTT-1_1, TCCTTCTCAGTTTCAG-1_1, TGTGAGTCAAATGATG-1_1, AAACGAAGTACAGAGC-1_1, CAACCAAAGTATTCCG-1_1, CTTCTCTCAGAGACTG-1_1, TACAACGGTGGCTGAA-1_1, AACGGGACATGCCGGT-1_1, AACCAACGTTGGGAAC-1_1, TATATCCAGCGTCAGA-1_1, AGACAAACATCCCGTT-1_1, ATGACCAGTCTTCATT-1_1, CTTACCGTCAGACATC-1_1, CGGGACTGTTAGTTCG-1_1, ATTCATCCACTGAGTT-1_1, TCATGAGAGAGGCGGA-1_1, TCCCACATCTAGTACG-1_1, CTTCCTTCATATCTCT-1_1, CCGGACACACTCGATA-1_1, ACACGCGCACCTGTCT-1_1, GAATCGTCAGAAGTGC-1_1, GGTGGCTCAAGCTCTA-1_1, CCTGCATCACATATGC-1_1, GTGGGAAGTTAAAGTG-1_1, CGTTCTGGTACTAGCT-1_1, ACCCTCACAATAGTCC-1_1, GCCCGAACAAACTAAG-1_1, GTGGAAGCACATGACT-1_1, GTTGTGACATCGATAC-1_1, GACAGCCCAGGTCCGT-1_1, TAAGCACGTTGGCTAT-1_1, GGGACAAGTCACCACG-1_1, CTGGCAGGTTCGGTAT-1_1, GACTCAACACTGTGAT-1_1, GCCAGTGGTGTGGTCC-1_1, TCTAACTGTAGGCAGT-1_1, GAAGAATGTAGCTTGT-1_1, TCACTCGCAATCTCTT-1_1, CAAGACTTCCCACAGG-1_1, CAGATACGTGACTCTA-1_1, TGGGATTAGAGGGTCT-1_1, ACCTGAACACTCCTTG-1_1, GACACGCCACTCGATA-1_1, CTCATCGTCACCGCTT-1_1, AGGTAGGGTCCCTGTT-1_1, TACCCGTCAACACTAC-1_1, TGCTGAAAGACGGATC-1_1, ACACCAACACAACGCC-1_1, AAGATAGCAAATGGAT-1_1, CTTCAATGTGACAGGT-1_1, GAAGCGAAGAGTTGAT-1_1, GCCGTGACACAAGCCC-1_1, CCTCAACCATACAGGG-1_1, ACAAAGATCCACAGGC-1_1, CAGATACAGTCCCAGC-1_1, GGCAGTCTCCGGTTCT-1_1, TAAGTCGAGCTGAGCA-1_1, GAGACCCGTCTGTGCG-1_1, TAACACGCATGTGTCA-1_1, TCAATTCGTTCTCGCT-1_1, GCTTTCGCACAGTGTT-1_1, AACCAACAGATAACAC-1_1, ATCGGCGCACATCATG-1_1, TCATCCGCACGAGGAT-1_1, CTGATCCTCTTTACAC-1_1, TCACACCCAACTTCTT-1_1, TGAGGGACACCGTACG-1_1, GTGCACGTCATCTGTT-1_1, GGTAATCAGTTGCATC-1_1, ATACTTCCAAGGTCTT-1_1, ...]

[6099 rows x 0 columns],
 &#39;obs_names&#39;: Index([&#39;TAGGTACCATGGCCAC-1_1&#39;, &#39;ATTCACTGTAACAGGC-1_1&#39;, &#39;TAACTTCCAACCACGC-1_1&#39;,
       &#39;TTGGGTACACGACAAG-1_1&#39;, &#39;AGGCCACAGAGTCACG-1_1&#39;, &#39;CACTGAAGTCGAAGCA-1_1&#39;,
       &#39;ACTGATGTCTGCACCT-1_1&#39;, &#39;TTACCGCCACTCAGAT-1_1&#39;, &#39;TTGGTTTTCCTAGCTC-1_1&#39;,
       &#39;TGGGAAGTCAGTGATC-1_1&#39;,
       ...
       &#39;AAGTCGTGTGTGAATA-1_1&#39;, &#39;GTCGTTCTCCAAGGGA-1_1&#39;, &#39;GTTTGGATCGGCCTTT-1_1&#39;,
       &#39;GTACAGTCACGTATAC-1_1&#39;, &#39;TCATGCCCAAGAGGTC-1_1&#39;, &#39;CGCCATTGTTTGCCGG-1_1&#39;,
       &#39;CACTGGGGTCTACGTA-1_1&#39;, &#39;CATACTTGTAGAGGAA-1_1&#39;, &#39;TTTGGTTTCCACGGAC-1_1&#39;,
       &#39;ATGCATGAGTCATGAA-1_1&#39;],
      dtype=&#39;object&#39;, length=6099),
 &#39;obsm&#39;: AxisArrays with keys: ,
 &#39;obsp&#39;: PairwiseArrays with keys: ,
 &#39;raw&#39;: None,
 &#39;shape&#39;: (6099, 34546),
 &#39;uns&#39;: OverloadedDict, wrapping:
        OrderedDict()
With overloaded keys:
        [&#39;neighbors&#39;].,
 &#39;var&#39;: Empty DataFrame
Columns: []
Index: [AL627309.1, AL627309.5, AL627309.4, AL669831.2, LINC01409, FAM87B, LINC01128, LINC00115, FAM41C, AL645608.6, AL645608.2, LINC02593, SAMD11, NOC2L, KLHL17, PLEKHN1, PERM1, AL645608.7, HES4, ISG15, AL645608.1, AGRN, C1orf159, AL390719.3, LINC01342, AL390719.2, TTLL10-AS1, TTLL10, TNFRSF18, TNFRSF4, SDF4, B3GALT6, C1QTNF12, UBE2J2, LINC01786, SCNN1D, ACAP3, PUSL1, INTS11, AL139287.1, CPTP, DVL1, MXRA8, AURKAIP1, CCNL2, MRPL20-AS1, MRPL20, AL391244.2, ANKRD65, AL391244.1, LINC01770, VWA1, ATAD3C, ATAD3B, ATAD3A, TMEM240, SSU72, AL645728.1, FNDC10, AL691432.4, AL691432.2, MIB2, MMP23B, CDK11B, FO704657.1, SLC35E2B, CDK11A, SLC35E2A, NADK, GNB1, AL109917.1, CALML6, TMEM52, CFAP74, GABRD, AL391845.1, PRKCZ, AL590822.2, PRKCZ-AS1, FAAP20, AL590822.1, SKI, AL590822.3, MORN1, AL589739.1, AL513477.2, RER1, PEX10, PLCH2, AL139246.4, PANK4, HES5, AL139246.5, TNFRSF14-AS1, TNFRSF14, AL139246.3, PRXL2B, MMEL1, TTC34, AC242022.2, ...]

[34546 rows x 0 columns],
 &#39;var_names&#39;: Index([&#39;AL627309.1&#39;, &#39;AL627309.5&#39;, &#39;AL627309.4&#39;, &#39;AL669831.2&#39;, &#39;LINC01409&#39;,
       &#39;FAM87B&#39;, &#39;LINC01128&#39;, &#39;LINC00115&#39;, &#39;FAM41C&#39;, &#39;AL645608.6&#39;,
       ...
       &#39;AC087190.2&#39;, &#39;AC136428.1&#39;, &#39;AC019183.1&#39;, &#39;AC105094.1&#39;, &#39;AC010485.1&#39;,
       &#39;VN1R2&#39;, &#39;AL031676.1&#39;, &#39;SMIM34A&#39;, &#39;AL050402.1&#39;, &#39;AL445072.1&#39;],
      dtype=&#39;object&#39;, length=34546),
 &#39;varm&#39;: AxisArrays with keys: ,
 &#39;varp&#39;: PairwiseArrays with keys: }
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_223/22660072.py:11: DeprecationWarning: Use is_view instead of isview, isview will be removed in the future.
  if name[0] != &#34;_&#34; and name not in disallowed_names and hasattr(obj, name)
/tmp/ipykernel_223/22660072.py:9: DeprecationWarning: Use is_view instead of isview, isview will be removed in the future.
  name: getattr(obj, name)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TAGGTACCATGGCCAC-1_1</th>
    </tr>
    <tr>
      <th>ATTCACTGTAACAGGC-1_1</th>
    </tr>
    <tr>
      <th>TAACTTCCAACCACGC-1_1</th>
    </tr>
    <tr>
      <th>TTGGGTACACGACAAG-1_1</th>
    </tr>
    <tr>
      <th>AGGCCACAGAGTCACG-1_1</th>
    </tr>
    <tr>
      <th>...</th>
    </tr>
    <tr>
      <th>CGCCATTGTTTGCCGG-1_1</th>
    </tr>
    <tr>
      <th>CACTGGGGTCTACGTA-1_1</th>
    </tr>
    <tr>
      <th>CATACTTGTAGAGGAA-1_1</th>
    </tr>
    <tr>
      <th>TTTGGTTTCCACGGAC-1_1</th>
    </tr>
    <tr>
      <th>ATGCATGAGTCATGAA-1_1</th>
    </tr>
  </tbody>
</table>
<p>6099 rows × 0 columns</p>
</div></div>
</div>
<p>Gene names are saved as <code class="docutils literal notranslate"><span class="pre">adata.var</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AL627309.1</th>
    </tr>
    <tr>
      <th>AL627309.5</th>
    </tr>
    <tr>
      <th>AL627309.4</th>
    </tr>
    <tr>
      <th>AL669831.2</th>
    </tr>
    <tr>
      <th>LINC01409</th>
    </tr>
    <tr>
      <th>...</th>
    </tr>
    <tr>
      <th>VN1R2</th>
    </tr>
    <tr>
      <th>AL031676.1</th>
    </tr>
    <tr>
      <th>SMIM34A</th>
    </tr>
    <tr>
      <th>AL050402.1</th>
    </tr>
    <tr>
      <th>AL445072.1</th>
    </tr>
  </tbody>
</table>
<p>34546 rows × 0 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;TAGGTACCATGGCCAC-1_1&#39;, &#39;ATTCACTGTAACAGGC-1_1&#39;, &#39;TAACTTCCAACCACGC-1_1&#39;,
       &#39;TTGGGTACACGACAAG-1_1&#39;, &#39;AGGCCACAGAGTCACG-1_1&#39;, &#39;CACTGAAGTCGAAGCA-1_1&#39;,
       &#39;ACTGATGTCTGCACCT-1_1&#39;, &#39;TTACCGCCACTCAGAT-1_1&#39;, &#39;TTGGTTTTCCTAGCTC-1_1&#39;,
       &#39;TGGGAAGTCAGTGATC-1_1&#39;,
       ...
       &#39;AAGTCGTGTGTGAATA-1_1&#39;, &#39;GTCGTTCTCCAAGGGA-1_1&#39;, &#39;GTTTGGATCGGCCTTT-1_1&#39;,
       &#39;GTACAGTCACGTATAC-1_1&#39;, &#39;TCATGCCCAAGAGGTC-1_1&#39;, &#39;CGCCATTGTTTGCCGG-1_1&#39;,
       &#39;CACTGGGGTCTACGTA-1_1&#39;, &#39;CATACTTGTAGAGGAA-1_1&#39;, &#39;TTTGGTTTCCACGGAC-1_1&#39;,
       &#39;ATGCATGAGTCATGAA-1_1&#39;],
      dtype=&#39;object&#39;, length=6099)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;AL627309.1&#39;, &#39;AL627309.5&#39;, &#39;AL627309.4&#39;, &#39;AL669831.2&#39;, &#39;LINC01409&#39;,
       &#39;FAM87B&#39;, &#39;LINC01128&#39;, &#39;LINC00115&#39;, &#39;FAM41C&#39;, &#39;AL645608.6&#39;,
       ...
       &#39;AC087190.2&#39;, &#39;AC136428.1&#39;, &#39;AC019183.1&#39;, &#39;AC105094.1&#39;, &#39;AC010485.1&#39;,
       &#39;VN1R2&#39;, &#39;AL031676.1&#39;, &#39;SMIM34A&#39;, &#39;AL050402.1&#39;, &#39;AL445072.1&#39;],
      dtype=&#39;object&#39;, length=34546)
</pre></div></div>
</div>
<p>There are no layers in this data set.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Layers with keys:
</pre></div></div>
</div>
<p>There are no multidimensional observations or variables.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">varp</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AxisArrays with keys:
AxisArrays with keys:
PairwiseArrays with keys:
PairwiseArrays with keys:
</pre></div></div>
</div>
<p>The data appears to contain reads mapped to 6099 cell-associated barcodes and 34546 RNA molecule-associated features.</p>
</div>
</div>
<div class="section" id="Doublet-removal">
<h2>Doublet removal<a class="headerlink" href="#Doublet-removal" title="Permalink to this heading">§</a></h2>
<div class="section" id="Filter-transcripts-by-minimum-number-of-cells-with-non-zero-counts">
<h3>Filter transcripts by minimum number of cells with non-zero counts<a class="headerlink" href="#Filter-transcripts-by-minimum-number-of-cells-with-non-zero-counts" title="Permalink to this heading">§</a></h3>
<p>We may choose to filter out transcript types that are detected in a relatively small number of cells. The optimum threshold is not known. Here we use 10 as a base case.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 34546
</pre></div></div>
</div>
<p>See the <a class="reference external" href="https://scanpy.readthedocs.io/en/latest/generated/scanpy.pp.filter_genes.html">documentation for scanpy pre-processing filter-genes</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 19896
    var: &#39;n_cells&#39;
</pre></div></div>
</div>
<p>Following filtration there are 19896 transcript types remaining that are present in at least 10 cells. This means 14650 transcript types were removed at the threshold of 10. Notice that an annotation named <code class="docutils literal notranslate"><span class="pre">n_cells</span></code> has been added to the genes to indicate the number of cells with non-zero values for that transcript type.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_cells</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AL627309.5</th>
      <td>33</td>
    </tr>
    <tr>
      <th>LINC01409</th>
      <td>274</td>
    </tr>
    <tr>
      <th>LINC01128</th>
      <td>81</td>
    </tr>
    <tr>
      <th>LINC00115</th>
      <td>15</td>
    </tr>
    <tr>
      <th>SAMD11</th>
      <td>16</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>MAFIP</th>
      <td>34</td>
    </tr>
    <tr>
      <th>AC011043.1</th>
      <td>11</td>
    </tr>
    <tr>
      <th>AL354822.1</th>
      <td>133</td>
    </tr>
    <tr>
      <th>AL592183.1</th>
      <td>1003</td>
    </tr>
    <tr>
      <th>AC240274.1</th>
      <td>162</td>
    </tr>
  </tbody>
</table>
<p>19896 rows × 1 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_cells</th>
      <th>highly_variable_rank</th>
      <th>means</th>
      <th>variances</th>
      <th>variances_norm</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>2000.000000</td>
      <td>2000.000000</td>
      <td>2000.000000</td>
      <td>2000.000000</td>
      <td>2000.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>319.636000</td>
      <td>999.500000</td>
      <td>0.140495</td>
      <td>1.135413</td>
      <td>2.203288</td>
    </tr>
    <tr>
      <th>std</th>
      <td>377.720811</td>
      <td>577.494568</td>
      <td>0.282680</td>
      <td>5.619025</td>
      <td>1.171426</td>
    </tr>
    <tr>
      <th>min</th>
      <td>10.000000</td>
      <td>0.000000</td>
      <td>0.001968</td>
      <td>0.002948</td>
      <td>1.399341</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>65.000000</td>
      <td>499.750000</td>
      <td>0.017216</td>
      <td>0.040875</td>
      <td>1.550249</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>184.500000</td>
      <td>999.500000</td>
      <td>0.053123</td>
      <td>0.148688</td>
      <td>1.828523</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>414.000000</td>
      <td>1499.250000</td>
      <td>0.148426</td>
      <td>0.545860</td>
      <td>2.346181</td>
    </tr>
    <tr>
      <th>max</th>
      <td>4546.000000</td>
      <td>1999.000000</td>
      <td>5.314150</td>
      <td>135.394535</td>
      <td>17.115198</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;n_cells&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:xlabel=&#39;n_cells&#39;, ylabel=&#39;Count&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/02-analysis-review_45_1.svg" src="_images/02-analysis-review_45_1.svg" /></div>
</div>
</div>
<div class="section" id="Select-highly-variable-genes">
<h3>Select highly variable genes<a class="headerlink" href="#Select-highly-variable-genes" title="Permalink to this heading">§</a></h3>
<p>It is common to select transcript types with high variability among the cell population under the assumption that this will help to focus on features that distinguish the cells. Again there is no perfect threshold. Here we select the 2000 highest variability genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 19896
    var: &#39;n_cells&#39;
</pre></div></div>
</div>
<p>See the <a class="reference external" href="https://scanpy.readthedocs.io/en/latest/generated/scanpy.pp.highly_variable_genes.html">documentation for scanpy pre-processing highly-variable-genes</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;seurat_v3&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 2000
    var: &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;
</pre></div></div>
</div>
<p>We have filtered down to 2000 transcript types and added 5 annotations to the genes including a binary variable indicating membership in the highly-variable class, the ranking among highly variable genes, and the mean, variance, and normalized variance for each gene across cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_cells</th>
      <th>highly_variable</th>
      <th>highly_variable_rank</th>
      <th>means</th>
      <th>variances</th>
      <th>variances_norm</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TTLL10</th>
      <td>112</td>
      <td>True</td>
      <td>903.0</td>
      <td>0.028857</td>
      <td>0.069354</td>
      <td>1.901045</td>
    </tr>
    <tr>
      <th>TNFRSF18</th>
      <td>15</td>
      <td>True</td>
      <td>1604.0</td>
      <td>0.002951</td>
      <td>0.004911</td>
      <td>1.513808</td>
    </tr>
    <tr>
      <th>CFAP74</th>
      <td>159</td>
      <td>True</td>
      <td>1370.0</td>
      <td>0.041154</td>
      <td>0.087024</td>
      <td>1.607399</td>
    </tr>
    <tr>
      <th>TTC34</th>
      <td>209</td>
      <td>True</td>
      <td>245.0</td>
      <td>0.080341</td>
      <td>0.363502</td>
      <td>3.044086</td>
    </tr>
    <tr>
      <th>AJAP1</th>
      <td>31</td>
      <td>True</td>
      <td>1922.0</td>
      <td>0.006886</td>
      <td>0.011432</td>
      <td>1.421002</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>MT-ND4L</th>
      <td>650</td>
      <td>True</td>
      <td>1212.0</td>
      <td>0.191671</td>
      <td>0.559353</td>
      <td>1.695637</td>
    </tr>
    <tr>
      <th>MT-ND4</th>
      <td>1328</td>
      <td>True</td>
      <td>178.0</td>
      <td>0.833087</td>
      <td>9.742224</td>
      <td>3.486984</td>
    </tr>
    <tr>
      <th>MT-ND5</th>
      <td>886</td>
      <td>True</td>
      <td>443.0</td>
      <td>0.332514</td>
      <td>1.669344</td>
      <td>2.430210</td>
    </tr>
    <tr>
      <th>MT-ND6</th>
      <td>821</td>
      <td>True</td>
      <td>123.0</td>
      <td>0.383178</td>
      <td>3.357087</td>
      <td>3.985056</td>
    </tr>
    <tr>
      <th>MT-CYB</th>
      <td>1295</td>
      <td>True</td>
      <td>414.0</td>
      <td>0.622397</td>
      <td>4.384615</td>
      <td>2.487058</td>
    </tr>
  </tbody>
</table>
<p>2000 rows × 6 columns</p>
</div></div>
</div>
</div>
<div class="section" id="id4">
<h3>Doublet removal<a class="headerlink" href="#id4" title="Permalink to this heading">§</a></h3>
<p>Here we <code class="docutils literal notranslate"><span class="pre">scvi-tools</span></code> model as <a class="reference external" href="https://docs.scvi-tools.org/en/stable/api/reference/scvi.external.SOLO.html">an interface</a> to <a class="reference external" href="https://github.com/calico/solo">solo</a> from <span id="id5">[<a class="reference internal" href="references.html#id2" title="Nicholas J Bernstein, Nicole L Fong, Irene Lam, Margaret A Roy, David G Hendrickson, and David R Kelley. Solo: doublet identification in Single-Cell RNA-Seq via Semi-Supervised deep learning. Cell Syst, 11(1):95–101.e5, July 2020. doi:10.1016/j.cels.2020.05.010.">BFL+20</a>]</span>. <code class="docutils literal notranslate"><span class="pre">scvi</span></code> adds some unstructured data to interface with <code class="docutils literal notranslate"><span class="pre">torch</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
OverloadedDict, wrapping:
        OrderedDict([(&#39;hvg&#39;, {&#39;flavor&#39;: &#39;seurat_v3&#39;})])
With overloaded keys:
        [&#39;neighbors&#39;].
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
OverloadedDict, wrapping:
        OrderedDict([(&#39;hvg&#39;, {&#39;flavor&#39;: &#39;seurat_v3&#39;}), (&#39;_scvi_uuid&#39;, &#39;290021be-0332-459c-9577-d34349c03ff4&#39;), (&#39;_scvi_manager_uuid&#39;, &#39;49d2f232-173d-4da9-bff8-dd83ec7e62ad&#39;)])
With overloaded keys:
        [&#39;neighbors&#39;].
</pre></div></div>
</div>
<p>Train the <code class="docutils literal notranslate"><span class="pre">scvi</span></code> model.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vae</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">vae</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epoch 400/400: 100% 400/400 [03:27&lt;00:00,  2.12it/s, loss=323, v_num=1]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
`Trainer.fit` stopped: `max_epochs=400` reached.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epoch 400/400: 100% 400/400 [03:27&lt;00:00,  1.92it/s, loss=323, v_num=1]
</pre></div></div>
</div>
<p>After training the model <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> now contains <code class="docutils literal notranslate"><span class="pre">_scvi_batch</span></code> and <code class="docutils literal notranslate"><span class="pre">scvi_labels</span></code> annotations.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 2000
    obs: &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>_scvi_batch</th>
      <th>_scvi_labels</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TAGGTACCATGGCCAC-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>ATTCACTGTAACAGGC-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>TAACTTCCAACCACGC-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>TTGGGTACACGACAAG-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>AGGCCACAGAGTCACG-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>CGCCATTGTTTGCCGG-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>CACTGGGGTCTACGTA-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>CATACTTGTAGAGGAA-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>TTTGGTTTCCACGGAC-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>ATGCATGAGTCATGAA-1_1</th>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>6099 rows × 2 columns</p>
</div></div>
</div>
<p>Train the <code class="docutils literal notranslate"><span class="pre">solo</span></code> model.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">solo</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">external</span><span class="o">.</span><span class="n">SOLO</span><span class="o">.</span><span class="n">from_scvi_model</span><span class="p">(</span><span class="n">vae</span><span class="p">)</span>
<span class="n">solo</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-blue-fg">INFO    </span> Creating doublets, preparing SOLO model.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/usr/lib/python3.10/site-packages/anndata/_core/anndata.py:1785: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  [AnnData(sparse.csr_matrix(a.shape), obs=a.obs) for a in all_adatas],
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epoch 207/400:  52% 207/400 [01:43&lt;01:36,  2.01it/s, loss=0.299, v_num=1]
Monitored metric validation_loss did not improve in the last 30 records. Best score: 0.295. Signaling Trainer to stop.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">(</span><span class="n">solo</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scvi.external.solo._model.SOLO
</pre></div></div>
</div>
<p>We can now use the <code class="docutils literal notranslate"><span class="pre">solo</span></code> model to add a prediction annotation for doublet or singlet to our data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">solo</span><span class="o">.</span><span class="n">predict</span><span class="p">()</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">solo</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">soft</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="p">])</span>

<span class="n">df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>doublet</th>
      <th>singlet</th>
      <th>prediction</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TAGGTACCATGGCCAC-1_1</th>
      <td>1.729540</td>
      <td>-0.992126</td>
      <td>doublet</td>
    </tr>
    <tr>
      <th>ATTCACTGTAACAGGC-1_1</th>
      <td>2.275840</td>
      <td>-1.639090</td>
      <td>doublet</td>
    </tr>
    <tr>
      <th>TAACTTCCAACCACGC-1_1</th>
      <td>0.735813</td>
      <td>-0.259502</td>
      <td>doublet</td>
    </tr>
    <tr>
      <th>TTGGGTACACGACAAG-1_1</th>
      <td>1.618703</td>
      <td>-0.680344</td>
      <td>doublet</td>
    </tr>
    <tr>
      <th>AGGCCACAGAGTCACG-1_1</th>
      <td>1.398426</td>
      <td>-0.673206</td>
      <td>doublet</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>CGCCATTGTTTGCCGG-1_1</th>
      <td>-0.691126</td>
      <td>1.362337</td>
      <td>singlet</td>
    </tr>
    <tr>
      <th>CACTGGGGTCTACGTA-1_1</th>
      <td>-2.632991</td>
      <td>2.468916</td>
      <td>singlet</td>
    </tr>
    <tr>
      <th>CATACTTGTAGAGGAA-1_1</th>
      <td>-2.506424</td>
      <td>2.567346</td>
      <td>singlet</td>
    </tr>
    <tr>
      <th>TTTGGTTTCCACGGAC-1_1</th>
      <td>-3.308819</td>
      <td>2.476630</td>
      <td>singlet</td>
    </tr>
    <tr>
      <th>ATGCATGAGTCATGAA-1_1</th>
      <td>-2.357370</td>
      <td>2.206591</td>
      <td>singlet</td>
    </tr>
  </tbody>
</table>
<p>6099 rows × 3 columns</p>
</div></div>
</div>
<p>Counting the predicted doublets and singlets reveals about 20% doublets.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;prediction&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>doublet</th>
      <th>singlet</th>
    </tr>
    <tr>
      <th>prediction</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>doublet</th>
      <td>1206</td>
      <td>1206</td>
    </tr>
    <tr>
      <th>singlet</th>
      <td>4893</td>
      <td>4893</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We can assess the magnitude of the prediction by taking the difference between the doublet and singlet scores.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;dif&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">doublet</span> <span class="o">-</span> <span class="n">df</span><span class="o">.</span><span class="n">singlet</span>
<span class="n">df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>doublet</th>
      <th>singlet</th>
      <th>prediction</th>
      <th>dif</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TAGGTACCATGGCCAC-1_1</th>
      <td>1.729540</td>
      <td>-0.992126</td>
      <td>doublet</td>
      <td>2.721666</td>
    </tr>
    <tr>
      <th>ATTCACTGTAACAGGC-1_1</th>
      <td>2.275840</td>
      <td>-1.639090</td>
      <td>doublet</td>
      <td>3.914930</td>
    </tr>
    <tr>
      <th>TAACTTCCAACCACGC-1_1</th>
      <td>0.735813</td>
      <td>-0.259502</td>
      <td>doublet</td>
      <td>0.995315</td>
    </tr>
    <tr>
      <th>TTGGGTACACGACAAG-1_1</th>
      <td>1.618703</td>
      <td>-0.680344</td>
      <td>doublet</td>
      <td>2.299047</td>
    </tr>
    <tr>
      <th>AGGCCACAGAGTCACG-1_1</th>
      <td>1.398426</td>
      <td>-0.673206</td>
      <td>doublet</td>
      <td>2.071632</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>CGCCATTGTTTGCCGG-1_1</th>
      <td>-0.691126</td>
      <td>1.362337</td>
      <td>singlet</td>
      <td>-2.053463</td>
    </tr>
    <tr>
      <th>CACTGGGGTCTACGTA-1_1</th>
      <td>-2.632991</td>
      <td>2.468916</td>
      <td>singlet</td>
      <td>-5.101907</td>
    </tr>
    <tr>
      <th>CATACTTGTAGAGGAA-1_1</th>
      <td>-2.506424</td>
      <td>2.567346</td>
      <td>singlet</td>
      <td>-5.073770</td>
    </tr>
    <tr>
      <th>TTTGGTTTCCACGGAC-1_1</th>
      <td>-3.308819</td>
      <td>2.476630</td>
      <td>singlet</td>
      <td>-5.785450</td>
    </tr>
    <tr>
      <th>ATGCATGAGTCATGAA-1_1</th>
      <td>-2.357370</td>
      <td>2.206591</td>
      <td>singlet</td>
      <td>-4.563961</td>
    </tr>
  </tbody>
</table>
<p>6099 rows × 4 columns</p>
</div></div>
</div>
<p>Plotting the distribution of doublet-singlet score differences we see there are a large fraction that marginally favor doublet to singlet.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">prediction</span> <span class="o">==</span> <span class="s2">&quot;doublet&quot;</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;dif&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:xlabel=&#39;dif&#39;, ylabel=&#39;Count&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/02-analysis-review_74_1.svg" src="_images/02-analysis-review_74_1.svg" /></div>
</div>
<p>Since we would like to avoid unnecessarily discarding barcodes, we will arbitrarily retain those with a score from zero to one (however this is not intended to be principled).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">doublets</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="o">.</span><span class="n">prediction</span> <span class="o">==</span> <span class="s1">&#39;doublet&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">dif</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)]</span>
<span class="n">doublets</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>doublet</th>
      <th>singlet</th>
      <th>prediction</th>
      <th>dif</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TAGGTACCATGGCCAC-1_1</th>
      <td>1.729540</td>
      <td>-0.992126</td>
      <td>doublet</td>
      <td>2.721666</td>
    </tr>
    <tr>
      <th>ATTCACTGTAACAGGC-1_1</th>
      <td>2.275840</td>
      <td>-1.639090</td>
      <td>doublet</td>
      <td>3.914930</td>
    </tr>
    <tr>
      <th>TTGGGTACACGACAAG-1_1</th>
      <td>1.618703</td>
      <td>-0.680344</td>
      <td>doublet</td>
      <td>2.299047</td>
    </tr>
    <tr>
      <th>AGGCCACAGAGTCACG-1_1</th>
      <td>1.398426</td>
      <td>-0.673206</td>
      <td>doublet</td>
      <td>2.071632</td>
    </tr>
    <tr>
      <th>CACTGAAGTCGAAGCA-1_1</th>
      <td>1.305717</td>
      <td>-0.544905</td>
      <td>doublet</td>
      <td>1.850623</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>CAATACGCAATGTGGG-1_1</th>
      <td>0.471522</td>
      <td>-0.923330</td>
      <td>doublet</td>
      <td>1.394853</td>
    </tr>
    <tr>
      <th>GGGTATTTCAGCGCAC-1_1</th>
      <td>1.101736</td>
      <td>-0.658175</td>
      <td>doublet</td>
      <td>1.759911</td>
    </tr>
    <tr>
      <th>TTGCTGCAGTGCGACA-1_1</th>
      <td>0.720216</td>
      <td>-0.624341</td>
      <td>doublet</td>
      <td>1.344557</td>
    </tr>
    <tr>
      <th>CATCCCAAGACGCCAA-1_1</th>
      <td>0.937599</td>
      <td>-0.486121</td>
      <td>doublet</td>
      <td>1.423720</td>
    </tr>
    <tr>
      <th>CACTGTCAGGGACAGG-1_1</th>
      <td>1.190305</td>
      <td>-0.678417</td>
      <td>doublet</td>
      <td>1.868721</td>
    </tr>
  </tbody>
</table>
<p>451 rows × 4 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">doublets</span><span class="p">[</span><span class="n">doublets</span><span class="o">.</span><span class="n">prediction</span> <span class="o">==</span> <span class="s2">&quot;doublet&quot;</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;dif&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:xlabel=&#39;dif&#39;, ylabel=&#39;Count&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/02-analysis-review_77_1.svg" src="_images/02-analysis-review_77_1.svg" /></div>
</div>
</div>
<div class="section" id="Save/load-checkpoint">
<h3>Save/load checkpoint<a class="headerlink" href="#Save/load-checkpoint" title="Permalink to this heading">§</a></h3>
<p>Save current variables to file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">([</span><span class="n">adata</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">doublets</span><span class="p">,</span> <span class="n">solo</span><span class="p">,</span> <span class="n">vae</span><span class="p">],</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;session_09282022.p&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<p>Variables can be reloaded if necessary.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">doublets</span><span class="p">,</span> <span class="n">solo</span><span class="p">,</span> <span class="n">vae</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;session_09282022.p&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Reload-data-and-filter-doublets">
<h3>Reload data and filter doublets<a class="headerlink" href="#Reload-data-and-filter-doublets" title="Permalink to this heading">§</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s2">&quot;data/GSE171524/supplementary/GSM5226574_C51ctr_raw_counts.csv.gz&quot;</span>
<span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 34546
</pre></div></div>
</div>
<p>Annotate <code class="docutils literal notranslate"><span class="pre">adata</span></code> with doublet column.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;doublet&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">doublets</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 6099 × 34546
    obs: &#39;doublet&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>doublet</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TAGGTACCATGGCCAC-1_1</th>
      <td>True</td>
    </tr>
    <tr>
      <th>ATTCACTGTAACAGGC-1_1</th>
      <td>True</td>
    </tr>
    <tr>
      <th>TAACTTCCAACCACGC-1_1</th>
      <td>False</td>
    </tr>
    <tr>
      <th>TTGGGTACACGACAAG-1_1</th>
      <td>True</td>
    </tr>
    <tr>
      <th>AGGCCACAGAGTCACG-1_1</th>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>CGCCATTGTTTGCCGG-1_1</th>
      <td>False</td>
    </tr>
    <tr>
      <th>CACTGGGGTCTACGTA-1_1</th>
      <td>False</td>
    </tr>
    <tr>
      <th>CATACTTGTAGAGGAA-1_1</th>
      <td>False</td>
    </tr>
    <tr>
      <th>TTTGGTTTCCACGGAC-1_1</th>
      <td>False</td>
    </tr>
    <tr>
      <th>ATGCATGAGTCATGAA-1_1</th>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>6099 rows × 1 columns</p>
</div></div>
</div>
<p>Use doublet column to filter doublets.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="o">~</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">doublet</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
View of AnnData object with n_obs × n_vars = 5648 × 34546
    obs: &#39;doublet&#39;
</pre></div></div>
</div>
<p>We filtered about 7% of the barcodes as putative doublets.</p>
</div>
</div>
<div class="section" id="Preprocessing">
<h2>Preprocessing<a class="headerlink" href="#Preprocessing" title="Permalink to this heading">§</a></h2>
<div class="section" id="Filter-mitochondrial-transcripts">
<h3>Filter mitochondrial transcripts<a class="headerlink" href="#Filter-mitochondrial-transcripts" title="Permalink to this heading">§</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<script type="application/vnd.jupyter.widget-state+json">
{"state": {}, "version_major": 2, "version_minor": 0}
</script></div>
</div>
</div>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="03-velocity-review.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Velocity review</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="01-data-structures-review.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Data structures review</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2022, Cameron Smith
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/cameronraysmith/sc" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Analysis review</a><ul>
<li><a class="reference internal" href="#Overview">Overview</a></li>
<li><a class="reference internal" href="#Setup">Setup</a><ul>
<li><a class="reference internal" href="#Import-libraries">Import libraries</a></li>
<li><a class="reference internal" href="#Setup-plotting">Setup plotting</a></li>
<li><a class="reference internal" href="#Utility-functions">Utility functions</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Import-data">Import data</a><ul>
<li><a class="reference internal" href="#Data-download">Data download</a></li>
<li><a class="reference internal" href="#Data-load">Data load</a></li>
<li><a class="reference internal" href="#Data-properties">Data properties</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Doublet-removal">Doublet removal</a><ul>
<li><a class="reference internal" href="#Filter-transcripts-by-minimum-number-of-cells-with-non-zero-counts">Filter transcripts by minimum number of cells with non-zero counts</a></li>
<li><a class="reference internal" href="#Select-highly-variable-genes">Select highly variable genes</a></li>
<li><a class="reference internal" href="#id4">Doublet removal</a></li>
<li><a class="reference internal" href="#Save/load-checkpoint">Save/load checkpoint</a></li>
<li><a class="reference internal" href="#Reload-data-and-filter-doublets">Reload data and filter doublets</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Preprocessing">Preprocessing</a><ul>
<li><a class="reference internal" href="#Filter-mitochondrial-transcripts">Filter mitochondrial transcripts</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/tabs.js"></script>
    <script src="_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "tags": "ams", "useLabelIds": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>