
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>RNA Velocity review &#8212; Single cell analysis</title>
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/insipid.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script defer src="_static/insipid.js"></script>
    <script defer src="_static/insipid-sidebar.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="References" href="references.html" />
    <link rel="prev" title="Basic single-cell analysis" href="02-analysis-review.html" />
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-capable" content="yes">
  </head><body>
    <script type="text/javascript">
        document.body.classList.add('js');
    </script>
    <input type="checkbox" id="sidebar-checkbox" style="display: none;">
    <label id="overlay" for="sidebar-checkbox"></label>
    <div class="sidebar-resize-handle"></div>
    <script type="text/javascript">
        try {
            let sidebar = localStorage.getItem('sphinx-sidebar');
            const sidebar_width = localStorage.getItem('sphinx-sidebar-width');
            if (sidebar_width) {
                document.documentElement.style.setProperty('--sidebar-width', sidebar_width);
            }
            // show sidebar on wide screen
            if (!sidebar && window.matchMedia('(min-width: 100rem)').matches) {
                sidebar = 'visible';
                // NB: We don't store the value in localStorage!
            }
            if (sidebar === 'visible') {
                document.getElementById('sidebar-checkbox').checked = true;
            }
        } catch(e) {
            console.info(e);
        }
    </script>
    <header id="topbar-placeholder">
      <div id="topbar">
        <div id="titlebar">
          <div class="buttons">
            <label for="sidebar-checkbox" id="sidebar-button" role="button" tabindex="0" aria-controls="sphinxsidebar" accesskey="M">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"/></svg>
            </label>
            <button id="search-button" type="button" title="Search" aria-label="Search" aria-expanded="false" aria-controls="search-form" accesskey="S">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"/></svg>
            </button>
          </div>
          <div class="title">
            <a class="parent" href="index.html" accesskey="U">Single cell analysis</a>
            <a class="top" href="#">RNA Velocity review</a>
          </div>
          <div class="buttons">
            <button id="fullscreen-button" type="button" aria-hidden="true">
              <span class="enable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M212.686 315.314L120 408l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C10.697 480 0 469.255 0 456V344c0-21.382 25.803-32.09 40.922-16.971L72 360l92.686-92.686c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.249 6.248 6.249 16.378 0 22.627zm22.628-118.628L328 104l-32.922-31.029C279.958 57.851 290.666 32 312.048 32h112C437.303 32 448 42.745 448 56v112c0 21.382-25.803 32.09-40.922 16.971L376 152l-92.686 92.686c-6.248 6.248-16.379 6.248-22.627 0l-25.373-25.373c-6.249-6.248-6.249-16.378 0-22.627z"/></svg>
              </span>
              <span class="disable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M4.686 427.314L104 328l-32.922-31.029C55.958 281.851 66.666 256 88.048 256h112C213.303 256 224 266.745 224 280v112c0 21.382-25.803 32.09-40.922 16.971L152 376l-99.314 99.314c-6.248 6.248-16.379 6.248-22.627 0L4.686 449.941c-6.248-6.248-6.248-16.379 0-22.627zM443.314 84.686L344 184l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C234.697 256 224 245.255 224 232V120c0-21.382 25.803-32.09 40.922-16.971L296 136l99.314-99.314c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.248 6.248 6.248 16.379 0 22.627z"/></svg>
              </span>
            </button>
          </div>
        </div>
        <div id="searchbox" role="search">
          <form id="search-form" class="search" style="display: none" action="search.html" method="get">
            <input type="search" name="q" placeholder="Search ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" />
            <button>Go</button>
          </form>
        </div>
      </div>
    </header>
    <nav>
      <a href="02-analysis-review.html" class="nav-icon previous" title="previous:&#13;Basic single-cell analysis" aria-label="Previous topic" accesskey="P" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 256 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"/></svg>
      </a>
      <a href="references.html" class="nav-icon next" title="next:&#13;References" aria-label="Next topic" accesskey="N" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 256 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"/></svg>
      </a>
    </nav>

    <nav class="relbar">
      <a class="previous" href="02-analysis-review.html" aria-label="Previous topic" >
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"/></svg>
        </div>
        <div class="title">
          <span class="text">
            <span class="direction">previous</span>
            Basic single-cell analysis
          </span>
        </div>
      </a>
      <a class="next" href="references.html" aria-label="Next topic" >
        <div class="title">
          <span class="text">
            <span class="direction">next</span>
            References
          </span>
        </div>
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"/></svg>
        </div>
      </a>
    </nav>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            

<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="admonition note">
  This page was generated from
  <a class="reference external" href="https://github.com/cameronraysmith/sc/blob/master/source/03-velocity-review.ipynb">source/03-velocity-review.ipynb</a>.
  <span style="white-space: nowrap;"><a href="https://mybinder.org/v2/gh/cameronraysmith/sc/master?filepath=source/03-velocity-review.ipynb"><img alt="Binder badge" src="https://mybinder.org/badge_logo.svg" style="vertical-align:text-bottom"></a> or </span>
  <script>
    if (document.location.host) {
      $(document.currentScript).replaceWith(
        '<a class="reference external" ' +
        'href="https://nbviewer.jupyter.org/url' +
        (window.location.protocol == 'https:' ? 's/' : '/') +
        window.location.host +
        window.location.pathname.slice(0, -4) +
        'ipynb"><em>nbviewer</em></a>.'
      );
    }
  </script>
</div><section id="RNA-Velocity-review">
<h1>RNA Velocity review<a class="headerlink" href="#RNA-Velocity-review" title="Permalink to this headline">¶</a></h1>
<p>Here we review basic RNA velocity analysis.</p>
<p>This is a modified version of the <a class="reference external" href="https://nbviewer.jupyter.org/github/theislab/scvelo_notebooks/blob/master/VelocityBasics.ipynb">scvelo basic velocity analysis</a> tutorial.</p>
<p>The primary data set is development of the endocrine pancreas, which demonstrates lineage commitment to four distinct fates: α, β, δ and ε-cells. See <a class="reference external" href="https://scvelo.readthedocs.io/scvelo.datasets.pancreas.html">here</a> for more details.</p>
<section id="Setup">
<h2>Setup<a class="headerlink" href="#Setup" title="Permalink to this headline">¶</a></h2>
<section id="Import-libraries">
<h3>Import libraries<a class="headerlink" href="#Import-libraries" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># update to the latest version, if not done yet.</span>
<span class="c1"># !pip install scvelo --upgrade --quiet</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">inspect</span> <span class="kn">import</span> <span class="n">getmembers</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>
<span class="kn">from</span> <span class="nn">types</span> <span class="kn">import</span> <span class="n">FunctionType</span>

<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scvelo</span> <span class="k">as</span> <span class="nn">scv</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_version</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running scvelo 0.2.4 (python 3.10.7) on 2022-09-26 17:41.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
ERROR: XMLRPC request failed [code: -32500]
RuntimeError: PyPI&#39;s XMLRPC API is currently disabled due to unmanageable load and will be deprecated in the near future. See https://status.python.org/ for more information.
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">3</span>  <span class="c1"># show errors(0), warnings(1), info(2), hints(3)</span>
<span class="n">scv</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">presenter_view</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># set max width size for presenter view</span>
<span class="n">scv</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="s1">&#39;scvelo&#39;</span><span class="p">)</span>  <span class="c1"># for beautified visualization</span>
<br/></pre></div>
</div>
</div>
</section>
<section id="Setup-plotting">
<h3>Setup plotting<a class="headerlink" href="#Setup-plotting" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.font_manager</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1"># import matplotlib_inline</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># fonts_path = &quot;/usr/share/texmf/fonts/opentype/public/lm/&quot; #ubuntu</span>
<span class="c1"># fonts_path = &quot;~/Library/Fonts/&quot; # macos</span>
<span class="n">fonts_path</span> <span class="o">=</span> <span class="s2">&quot;/usr/share/fonts/OTF/&quot;</span>  <span class="c1"># arch</span>
<span class="c1"># user_path = &quot;$HOME/&quot; # user</span>
<span class="c1"># fonts_path = user_path + &quot;fonts/latinmodern/opentype/public/lm/&quot;  # home</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">font_manager</span><span class="o">.</span><span class="n">fontManager</span><span class="o">.</span><span class="n">addfont</span><span class="p">(</span><span class="n">fonts_path</span> <span class="o">+</span> <span class="s2">&quot;lmsans10-regular.otf&quot;</span><span class="p">)</span>
<span class="n">matplotlib</span><span class="o">.</span><span class="n">font_manager</span><span class="o">.</span><span class="n">fontManager</span><span class="o">.</span><span class="n">addfont</span><span class="p">(</span><span class="n">fonts_path</span> <span class="o">+</span> <span class="s2">&quot;lmroman10-regular.otf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># https://stackoverflow.com/a/36622238/446907</span>
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_formats = [&#39;svg&#39;]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s2">&quot;default&quot;</span><span class="p">)</span>  <span class="c1"># reset default parameters</span>
<span class="c1"># https://stackoverflow.com/a/3900167/446907</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;font.size&quot;</span><span class="p">:</span> <span class="mi">16</span><span class="p">,</span>
        <span class="s2">&quot;font.family&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sans-serif&quot;</span><span class="p">],</span>
        <span class="s2">&quot;font.serif&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Latin Modern Roman&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;font.serif&quot;</span><span class="p">],</span>
        <span class="s2">&quot;font.sans-serif&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Latin Modern Sans&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;font.sans-serif&quot;</span><span class="p">],</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Utility-functions">
<h3>Utility functions<a class="headerlink" href="#Utility-functions" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">attributes</span><span class="p">(</span><span class="n">obj</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    get object attributes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">disallowed_names</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">name</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">getmembers</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">obj</span><span class="p">))</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">FunctionType</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="n">name</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;_&quot;</span> <span class="ow">and</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">disallowed_names</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
    <span class="p">}</span>


<span class="k">def</span> <span class="nf">print_attributes</span><span class="p">(</span><span class="n">obj</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    print object attributes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pprint</span><span class="p">(</span><span class="n">attributes</span><span class="p">(</span><span class="n">obj</span><span class="p">))</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>load an example dataset into an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object</p>
<ul>
<li><p>the data set used here contains precomputed spliced/unspliced reads</p></li>
<li><p>spliced/unspliced reads can be computed with <code class="docutils literal notranslate"><span class="pre">velocyto.py</span></code> and <code class="docutils literal notranslate"><span class="pre">kb-python</span></code></p></li>
</ul>
</li>
<li><p>view the detailed structure of the object</p></li>
<li><p>estimate relative proportions of spliced/unspliced reads in the data</p></li>
</ul>
<section id="Load-AnnData-object">
<h3>Load AnnData object<a class="headerlink" href="#Load-AnnData-object" title="Permalink to this headline">¶</a></h3>
<p>The analysis is based on the in-built <a class="reference external" href="https://scvelo.readthedocs.io/scvelo.datasets.pancreas">pancreas data</a>. To run velocity analysis on your own data, read your file (loom, h5ad, csv …) to an AnnData object with <code class="docutils literal notranslate"><span class="pre">adata</span> <span class="pre">=</span> <span class="pre">scv.read('path/file.loom',</span> <span class="pre">cache=True)</span></code>. If you want to merge your loom file into an already existing AnnData object, use <code class="docutils literal notranslate"><span class="pre">scv.utils.merge(adata,</span> <span class="pre">adata_loom)</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pancreas</span><span class="p">()</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 3696 × 27998
    obs: &#39;clusters_coarse&#39;, &#39;clusters&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;
    var: &#39;highly_variable_genes&#39;
    uns: &#39;clusters_coarse_colors&#39;, &#39;clusters_colors&#39;, &#39;day_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;spliced&#39;, &#39;unspliced&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
<p>scVelo is based on <code class="docutils literal notranslate"><span class="pre">adata</span></code>, an object that stores a data matrix <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>, annotation of observations <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>, variables <code class="docutils literal notranslate"><span class="pre">adata.var</span></code>, and unstructured annotations <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code>. Names of observations and variables can be accessed via <code class="docutils literal notranslate"><span class="pre">adata.obs_names</span></code> and <code class="docutils literal notranslate"><span class="pre">adata.var_names</span></code>, respectively. AnnData objects can be sliced like dataframes, for example, <code class="docutils literal notranslate"><span class="pre">adata_subset</span> <span class="pre">=</span> <span class="pre">adata[:,</span> <span class="pre">list_of_gene_names]</span></code>. For more details, see the <a class="reference external" href="https://anndata.readthedocs.io">anndata docs</a>.</p>
</section>
<section id="Detailed-view-of-adata">
<h3>Detailed view of adata<a class="headerlink" href="#Detailed-view-of-adata" title="Permalink to this headline">¶</a></h3>
<p>Observations are indexed by barcodes that are intended to correspond to cells after preprocessing. There are four components to the annotation vector. Two of these are fine- and coarse-grained clusters into cell types. The others are scores <code class="docutils literal notranslate"><span class="pre">S_score</span></code> and <code class="docutils literal notranslate"><span class="pre">G2M_score</span></code> that assess expression of G2/M and S phase marker genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clusters_coarse</th>
      <th>clusters</th>
      <th>S_score</th>
      <th>G2M_score</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCTGAGAGGGATA</th>
      <td>Pre-endocrine</td>
      <td>Pre-endocrine</td>
      <td>-0.224902</td>
      <td>-0.252071</td>
    </tr>
    <tr>
      <th>AAACCTGAGCCTTGAT</th>
      <td>Ductal</td>
      <td>Ductal</td>
      <td>-0.014707</td>
      <td>-0.232610</td>
    </tr>
    <tr>
      <th>AAACCTGAGGCAATTA</th>
      <td>Endocrine</td>
      <td>Alpha</td>
      <td>-0.171255</td>
      <td>-0.286834</td>
    </tr>
    <tr>
      <th>AAACCTGCATCATCCC</th>
      <td>Ductal</td>
      <td>Ductal</td>
      <td>0.599244</td>
      <td>0.191243</td>
    </tr>
    <tr>
      <th>AAACCTGGTAAGTGGC</th>
      <td>Ngn3 high EP</td>
      <td>Ngn3 high EP</td>
      <td>-0.179981</td>
      <td>-0.126030</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTTGTCAAGTGACATA</th>
      <td>Pre-endocrine</td>
      <td>Pre-endocrine</td>
      <td>-0.235896</td>
      <td>-0.266101</td>
    </tr>
    <tr>
      <th>TTTGTCAAGTGTGGCA</th>
      <td>Ngn3 high EP</td>
      <td>Ngn3 high EP</td>
      <td>0.279374</td>
      <td>-0.204047</td>
    </tr>
    <tr>
      <th>TTTGTCAGTTGTTTGG</th>
      <td>Ductal</td>
      <td>Ductal</td>
      <td>-0.045692</td>
      <td>-0.208907</td>
    </tr>
    <tr>
      <th>TTTGTCATCGAATGCT</th>
      <td>Endocrine</td>
      <td>Alpha</td>
      <td>-0.240576</td>
      <td>-0.206865</td>
    </tr>
    <tr>
      <th>TTTGTCATCTGTTTGT</th>
      <td>Endocrine</td>
      <td>Epsilon</td>
      <td>-0.136407</td>
      <td>-0.184763</td>
    </tr>
  </tbody>
</table>
<p>3696 rows × 4 columns</p>
</div></div>
</div>
<p>Gene names are saved</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>highly_variable_genes</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Xkr4</th>
      <td>False</td>
    </tr>
    <tr>
      <th>Gm37381</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Rp1</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Rp1-1</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Sox17</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>Gm28672</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Gm28670</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Gm29504</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Gm20837</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Erdr1</th>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>27998 rows × 1 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;AAACCTGAGAGGGATA&#39;, &#39;AAACCTGAGCCTTGAT&#39;, &#39;AAACCTGAGGCAATTA&#39;,
       &#39;AAACCTGCATCATCCC&#39;, &#39;AAACCTGGTAAGTGGC&#39;, &#39;AAACCTGGTATTAGCC&#39;,
       &#39;AAACCTGTCCCTCTTT&#39;, &#39;AAACCTGTCTTTCCTC&#39;, &#39;AAACGGGAGACAATAC&#39;,
       &#39;AAACGGGAGATATGGT&#39;,
       ...
       &#39;TTTGGTTCACCAGATT&#39;, &#39;TTTGGTTCACGAAGCA&#39;, &#39;TTTGGTTTCACTTACT&#39;,
       &#39;TTTGGTTTCCTTTCGG&#39;, &#39;TTTGTCAAGAATGTGT&#39;, &#39;TTTGTCAAGTGACATA&#39;,
       &#39;TTTGTCAAGTGTGGCA&#39;, &#39;TTTGTCAGTTGTTTGG&#39;, &#39;TTTGTCATCGAATGCT&#39;,
       &#39;TTTGTCATCTGTTTGT&#39;],
      dtype=&#39;object&#39;, name=&#39;index&#39;, length=3696)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;Xkr4&#39;, &#39;Gm37381&#39;, &#39;Rp1&#39;, &#39;Rp1-1&#39;, &#39;Sox17&#39;, &#39;Gm37323&#39;, &#39;Mrpl15&#39;,
       &#39;Rgs20&#39;, &#39;Npbwr1&#39;, &#39;4732440D04Rik&#39;,
       ...
       &#39;Gm28406&#39;, &#39;Gm29436&#39;, &#39;Gm28407&#39;, &#39;Gm29393&#39;, &#39;Gm21294&#39;, &#39;Gm28672&#39;,
       &#39;Gm28670&#39;, &#39;Gm29504&#39;, &#39;Gm20837&#39;, &#39;Erdr1&#39;],
      dtype=&#39;object&#39;, name=&#39;index&#39;, length=27998)
</pre></div></div>
</div>
<p>There are two layers corresponding to spliced and unspliced transcripts respectively.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;spliced&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;3696x27998 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
        with 9298890 stored elements in Compressed Sparse Row format&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;unspliced&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;3696x27998 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
        with 3156504 stored elements in Compressed Sparse Row format&gt;
</pre></div></div>
</div>
<p>PCA and UMAP have retained 50 and 2 dimensions respectively.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AxisArrays with keys: X_pca, X_umap
(3696, 50)
AxisArrays with keys: X_pca, X_umap
(3696, 2)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AxisArrays with keys:
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;distances&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;connectivities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PairwiseArrays with keys: distances, connectivities
(3696, 3696)
PairwiseArrays with keys: distances, connectivities
(3696, 3696)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">varp</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PairwiseArrays with keys:
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_attributes</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;T&#39;: AnnData object with n_obs × n_vars = 27998 × 3696
    obs: &#39;highly_variable_genes&#39;
    var: &#39;clusters_coarse&#39;, &#39;clusters&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;
    uns: &#39;clusters_coarse_colors&#39;, &#39;clusters_colors&#39;, &#39;day_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;
    varm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;spliced&#39;, &#39;unspliced&#39;
    varp: &#39;distances&#39;, &#39;connectivities&#39;,
 &#39;X&#39;: &lt;3696x27998 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
        with 9298890 stored elements in Compressed Sparse Row format&gt;,
 &#39;file&#39;: Backing file manager: no file is set.,
 &#39;filename&#39;: None,
 &#39;is_view&#39;: False,
 &#39;isbacked&#39;: False,
 &#39;isview&#39;: False,
 &#39;layers&#39;: Layers with keys: spliced, unspliced,
 &#39;n_obs&#39;: 3696,
 &#39;n_vars&#39;: 27998,
 &#39;obs&#39;:                  clusters_coarse       clusters   S_score  G2M_score
index
AAACCTGAGAGGGATA   Pre-endocrine  Pre-endocrine -0.224902  -0.252071
AAACCTGAGCCTTGAT          Ductal         Ductal -0.014707  -0.232610
AAACCTGAGGCAATTA       Endocrine          Alpha -0.171255  -0.286834
AAACCTGCATCATCCC          Ductal         Ductal  0.599244   0.191243
AAACCTGGTAAGTGGC    Ngn3 high EP   Ngn3 high EP -0.179981  -0.126030
...                          ...            ...       ...        ...
TTTGTCAAGTGACATA   Pre-endocrine  Pre-endocrine -0.235896  -0.266101
TTTGTCAAGTGTGGCA    Ngn3 high EP   Ngn3 high EP  0.279374  -0.204047
TTTGTCAGTTGTTTGG          Ductal         Ductal -0.045692  -0.208907
TTTGTCATCGAATGCT       Endocrine          Alpha -0.240576  -0.206865
TTTGTCATCTGTTTGT       Endocrine        Epsilon -0.136407  -0.184763

[3696 rows x 4 columns],
 &#39;obs_names&#39;: Index([&#39;AAACCTGAGAGGGATA&#39;, &#39;AAACCTGAGCCTTGAT&#39;, &#39;AAACCTGAGGCAATTA&#39;,
       &#39;AAACCTGCATCATCCC&#39;, &#39;AAACCTGGTAAGTGGC&#39;, &#39;AAACCTGGTATTAGCC&#39;,
       &#39;AAACCTGTCCCTCTTT&#39;, &#39;AAACCTGTCTTTCCTC&#39;, &#39;AAACGGGAGACAATAC&#39;,
       &#39;AAACGGGAGATATGGT&#39;,
       ...
       &#39;TTTGGTTCACCAGATT&#39;, &#39;TTTGGTTCACGAAGCA&#39;, &#39;TTTGGTTTCACTTACT&#39;,
       &#39;TTTGGTTTCCTTTCGG&#39;, &#39;TTTGTCAAGAATGTGT&#39;, &#39;TTTGTCAAGTGACATA&#39;,
       &#39;TTTGTCAAGTGTGGCA&#39;, &#39;TTTGTCAGTTGTTTGG&#39;, &#39;TTTGTCATCGAATGCT&#39;,
       &#39;TTTGTCATCTGTTTGT&#39;],
      dtype=&#39;object&#39;, name=&#39;index&#39;, length=3696),
 &#39;obsm&#39;: AxisArrays with keys: X_pca, X_umap,
 &#39;obsp&#39;: PairwiseArrays with keys: distances, connectivities,
 &#39;raw&#39;: None,
 &#39;shape&#39;: (3696, 27998),
 &#39;uns&#39;: OverloadedDict, wrapping:
        {&#39;clusters_coarse_colors&#39;: array([&#39;#031A5C&#39;, &#39;#fa9fb5&#39;, &#39;#fdbf6f&#39;, &#39;#ff7f00&#39;, &#39;#b2df8a&#39;],
      dtype=object), &#39;clusters_colors&#39;: array([&#39;#8fbc8f&#39;, &#39;#f4a460&#39;, &#39;#fdbf6f&#39;, &#39;#ff7f00&#39;, &#39;#b2df8a&#39;, &#39;#1f78b4&#39;,
       &#39;#6a3d9a&#39;, &#39;#cab2d6&#39;], dtype=object), &#39;day_colors&#39;: array([&#39;#d62728&#39;], dtype=object), &#39;neighbors&#39;: {&#39;params&#39;: {&#39;method&#39;: array([&#39;umap&#39;], dtype=object), &#39;n_neighbors&#39;: array([15])}}, &#39;pca&#39;: {&#39;variance&#39;: array([84.01987   , 56.14541   , 24.902763  , 18.046164  , 13.804973  ,
       11.931583  ,  9.848511  ,  5.831464  ,  4.772083  ,  4.364034  ,
        3.9239557 ,  3.4221544 ,  2.7514822 ,  2.4077806 ,  2.2492352 ,
        2.1247294 ,  2.067351  ,  1.7702986 ,  1.7164644 ,  1.5288209 ,
        1.4669604 ,  1.3977634 ,  1.341725  ,  1.2898475 ,  1.1913112 ,
        1.1588011 ,  1.1220304 ,  1.087722  ,  1.0535891 ,  0.999956  ,
        0.9759477 ,  0.96313083,  0.92936826,  0.91443396,  0.89059216,
        0.8624986 ,  0.85675824,  0.807511  ,  0.79329145,  0.779374  ,
        0.7724183 ,  0.7326805 ,  0.7189729 ,  0.69933045,  0.67954636,
        0.67046547,  0.63618135,  0.63056374,  0.6286348 ,  0.6084896 ],
      dtype=float32), &#39;variance_ratio&#39;: array([0.12779653, 0.08539871, 0.03787779, 0.02744871, 0.02099774,
       0.01814827, 0.01497986, 0.00886982, 0.00725847, 0.00663782,
       0.00596844, 0.00520519, 0.00418508, 0.0036623 , 0.00342115,
       0.00323177, 0.0031445 , 0.00269267, 0.00261079, 0.00232538,
       0.00223129, 0.00212604, 0.0020408 , 0.00196189, 0.00181202,
       0.00176257, 0.00170664, 0.00165445, 0.00160254, 0.00152096,
       0.00148444, 0.00146495, 0.00141359, 0.00139088, 0.00135462,
       0.00131188, 0.00130315, 0.00122825, 0.00120662, 0.00118545,
       0.00117487, 0.00111443, 0.00109358, 0.0010637 , 0.00103361,
       0.0010198 , 0.00096765, 0.0009591 , 0.00095617, 0.00092553],
      dtype=float32)}}
With overloaded keys:
        [&#39;neighbors&#39;].,
 &#39;var&#39;:         highly_variable_genes
index
Xkr4                    False
Gm37381                   NaN
Rp1                       NaN
Rp1-1                     NaN
Sox17                     NaN
...                       ...
Gm28672                   NaN
Gm28670                   NaN
Gm29504                   NaN
Gm20837                   NaN
Erdr1                    True

[27998 rows x 1 columns],
 &#39;var_names&#39;: Index([&#39;Xkr4&#39;, &#39;Gm37381&#39;, &#39;Rp1&#39;, &#39;Rp1-1&#39;, &#39;Sox17&#39;, &#39;Gm37323&#39;, &#39;Mrpl15&#39;,
       &#39;Rgs20&#39;, &#39;Npbwr1&#39;, &#39;4732440D04Rik&#39;,
       ...
       &#39;Gm28406&#39;, &#39;Gm29436&#39;, &#39;Gm28407&#39;, &#39;Gm29393&#39;, &#39;Gm21294&#39;, &#39;Gm28672&#39;,
       &#39;Gm28670&#39;, &#39;Gm29504&#39;, &#39;Gm20837&#39;, &#39;Erdr1&#39;],
      dtype=&#39;object&#39;, name=&#39;index&#39;, length=27998),
 &#39;varm&#39;: AxisArrays with keys: ,
 &#39;varp&#39;: PairwiseArrays with keys: }
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_140/22660072.py:11: DeprecationWarning: Use is_view instead of isview, isview will be removed in the future.
  if name[0] != &#34;_&#34; and name not in disallowed_names and hasattr(obj, name)
/tmp/ipykernel_140/22660072.py:9: DeprecationWarning: Use is_view instead of isview, isview will be removed in the future.
  name: getattr(obj, name)
</pre></div></div>
</div>
</section>
<section id="Show-spliced/unspliced">
<h3>Show spliced/unspliced<a class="headerlink" href="#Show-spliced/unspliced" title="Permalink to this headline">¶</a></h3>
<p>Here, the proportions of spliced/unspliced counts are displayed. Depending on the protocol used (Drop-Seq, Smart-Seq), we typically have between 10%-25% of unspliced molecules containing intronic sequences. We also advice you to examine the variations on cluster level to verify consistency in splicing efficiency. Here, we find variations as expected, with slightly lower unspliced proportions at cycling ductal cells, then higher proportion at cell fate commitment in Ngn3-high and Pre-endocrine
cells where many genes start to be transcribed.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">proportions</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_39_0.png" src="_images/03-velocity-review_39_0.png" />
</div>
</div>
</section>
</section>
<section id="Preprocess-the-Data">
<h2>Preprocess the Data<a class="headerlink" href="#Preprocess-the-Data" title="Permalink to this headline">¶</a></h2>
<p>Preprocessing requisites consist of <strong>gene selection</strong> by detection (with a minimum number of counts) and high variability (dispersion), <strong>normalizing</strong> every cell by its total size and <strong>logarithmizing</strong> X. Filtering and normalization is applied in the same vein to spliced/unspliced counts and X. Logarithmizing is only applied to X. If X is already preprocessed from former analysis, it will not be touched.</p>
<p>All of this is summarized in a single function <code class="docutils literal notranslate"><span class="pre">scv.pp.filter_and_normalize</span></code>, which essentially runs the following:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>scv.pp.filter_genes(adata, min_shared_counts=20)
scv.pp.normalize_per_cell(adata)
scv.pp.filter_genes_dispersion(adata, n_top_genes=2000)
scv.pp.log1p(adata)
</pre></div>
</div>
<p>Further, we need the first and second order moments (means and uncentered variances) computed among nearest neighbors in PCA space, summarized in <code class="docutils literal notranslate"><span class="pre">scv.pp.moments</span></code>, which internally computes <code class="docutils literal notranslate"><span class="pre">scv.pp.pca</span></code> and <code class="docutils literal notranslate"><span class="pre">scv.pp.neighbors</span></code>. First order is needed for deterministic velocity estimation, while stochastic estimation also requires second order moments.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_and_normalize</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_shared_counts</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">moments</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Filtered out 20801 genes that are detected in less than 20 counts (shared).
Normalized count data: X, spliced, unspliced.
Logarithmized X.
computing neighbors
    finished (0:00:03) --&gt; added
    &#39;distances&#39; and &#39;connectivities&#39;, weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:00) --&gt; added
    &#39;Ms&#39; and &#39;Mu&#39;, moments of spliced/unspliced abundances (adata.layers)
</pre></div></div>
</div>
<p>Further preprocessing (such as batch effect correction) may be used to remove unwanted sources of variability. See the <a class="reference external" href="https://doi.org/10.15252/msb.20188746">best practices</a> for further details. Note, that any additional preprocessing step only affects X and is not applied to spliced/unspliced counts.</p>
</section>
<section id="Compute-RNA-velocity">
<h2>Compute RNA velocity<a class="headerlink" href="#Compute-RNA-velocity" title="Permalink to this headline">¶</a></h2>
<section id="Estimate-RNA-velocity">
<h3>Estimate RNA velocity<a class="headerlink" href="#Estimate-RNA-velocity" title="Permalink to this headline">¶</a></h3>
<p>Velocities are vectors in gene expression space and represent the direction and speed of movement of the individual cells. The velocities are obtained by modeling transcriptional dynamics of splicing kinetics, either stochastically (default) or deterministically (by setting <code class="docutils literal notranslate"><span class="pre">mode='deterministic'</span></code>). For each gene, a steady-state-ratio of pre-mature (unspliced) and mature (spliced) mRNA counts is fitted, which constitutes a constant transcriptional state. Velocities are then obtained as
residuals from this ratio. Positive velocity indicates that a gene is up-regulated, which occurs for cells that show higher abundance of unspliced mRNA for that gene than expected in steady state. Conversely, negative velocity indicates that a gene is down-regulated.</p>
<p>The solution to the full dynamical model is obtained by setting <code class="docutils literal notranslate"><span class="pre">mode='dynamical'</span></code>, which requires to run <code class="docutils literal notranslate"><span class="pre">scv.tl.recover_dynamics(adata)</span></code> beforehand. We will elaborate more on the dynamical model in the next tutorial.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">velocity</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
computing velocities
    finished (0:00:01) --&gt; added
    &#39;velocity&#39;, velocity vectors for each individual cell (adata.layers)
</pre></div></div>
</div>
<p>The computed velocities are stored in <code class="docutils literal notranslate"><span class="pre">adata.layers</span></code> just like the count matrices.</p>
<p>The combination of velocities across genes can then be used to estimate the future state of an individual cell. In order to project the velocities into a lower-dimensional embedding, transition probabilities of cell-to-cell transitions are estimated. That is, for each velocity vector we find the likely cell transitions that are accordance with that direction. The transition probabilities are computed using cosine correlation between the potential cell-to-cell transitions and the velocity vector,
and are stored in a matrix denoted as velocity graph. The resulting velocity graph has dimension <span class="math notranslate nohighlight">\(n_{obs} \times n_{obs}\)</span> and summarizes the possible cell state changes that are well explained through the velocity vectors (for runtime speedup it can also be computed on reduced PCA space by setting <code class="docutils literal notranslate"><span class="pre">approx=True</span></code>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">velocity_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
computing velocity graph
    finished (0:00:10) --&gt; added
    &#39;velocity_graph&#39;, sparse matrix with cosine correlations (adata.uns)
</pre></div></div>
</div>
<p>For a variety of applications, the velocity graph can be converted to a transition matrix by applying a Gaussian kernel to transform the cosine correlations into actual transition probabilities. You can access the Markov transition matrix via <code class="docutils literal notranslate"><span class="pre">scv.utils.get_transition_matrix</span></code>.</p>
<p>As mentioned, it is internally used to project the velocities into a low-dimensional embedding by applying the mean transition with respect to the transition probabilities, obtained with <code class="docutils literal notranslate"><span class="pre">scv.tl.velocity_embedding</span></code>. Further, we can trace cells along the Markov chain to their origins and potential fates, thereby getting root cells and end points within a trajectory, obtained via <code class="docutils literal notranslate"><span class="pre">scv.tl.terminal_states</span></code>.</p>
</section>
<section id="Project-the-velocities">
<h3>Project the velocities<a class="headerlink" href="#Project-the-velocities" title="Permalink to this headline">¶</a></h3>
<p>Finally, the velocities are projected onto any embedding, specified by <code class="docutils literal notranslate"><span class="pre">basis</span></code>, and visualized in one of these ways: - on cellular level with <code class="docutils literal notranslate"><span class="pre">scv.pl.velocity_embedding</span></code>, - as gridlines with <code class="docutils literal notranslate"><span class="pre">scv.pl.velocity_embedding_grid</span></code>, - or as streamlines with <code class="docutils literal notranslate"><span class="pre">scv.pl.velocity_embedding_stream</span></code>.</p>
<p>Note, that the data has an already pre-computed UMAP embedding, and annotated clusters. When applying to your own data, these can be obtained with <code class="docutils literal notranslate"><span class="pre">scv.tl.umap</span></code> and <code class="docutils literal notranslate"><span class="pre">scv.tl.louvain</span></code>. For more details, see the <a class="reference external" href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">scanpy tutorial</a>. Further, all plotting functions are defaulted to using <code class="docutils literal notranslate"><span class="pre">basis='umap'</span></code> and <code class="docutils literal notranslate"><span class="pre">color='clusters'</span></code>, which you can set accordingly.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">velocity_embedding_stream</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
computing velocity embedding
    finished (0:00:00) --&gt; added
    &#39;velocity_umap&#39;, embedded velocity vectors (adata.obsm)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_53_1.png" src="_images/03-velocity-review_53_1.png" />
</div>
</div>
<p>The velocity vector field displayed as streamlines yields fine-grained insights into the developmental processes. It accurately delineates the cycling population of ductal cells and endocrine progenitors. Further, it illuminates cell states of lineage commitment, cell-cycle exit, and endocrine cell differentiation.</p>
<p>The most fine-grained resolution of the velocity vector field we get at single-cell level, with each arrow showing the direction and speed of movement of an individual cell. That reveals, e.g., the early endocrine commitment of Ngn3-cells (yellow) and a clear-cut difference between near-terminal α-cells (blue) and transient β-cells (green).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">velocity_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">arrow_length</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">arrow_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">120</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_56_0.png" src="_images/03-velocity-review_56_0.png" />
</div>
</div>
</section>
<section id="Interpret-the-velocities">
<h3>Interpret the velocities<a class="headerlink" href="#Interpret-the-velocities" title="Permalink to this headline">¶</a></h3>
<p>This is perhaps the most important part as we advise the user not to limit biological conclusions to the projected velocities, but to examine individual gene dynamics via phase portraits to understand how inferred directions are supported by particular genes.</p>
<p>See the gif <a class="reference external" href="https://user-images.githubusercontent.com/31883718/80227452-eb822480-864d-11ea-9399-56886c5e2785.gif">here</a> to get an idea of how to interpret a spliced vs. unspliced phase portrait. Gene activity is orchestrated by transcriptional regulation. Transcriptional induction for a particular gene results in an increase of (newly transcribed) precursor unspliced mRNAs while, conversely, repression or absence of transcription results in a decrease of unspliced mRNAs. Spliced mRNAs is
produced from unspliced mRNA and follows the same trend with a time lag. Time is a hidden/latent variable. Thus, the dynamics needs to be inferred from what is actually measured: spliced and unspliced mRNAs as displayed in the phase portrait.</p>
<p>Now, let us examine the phase portraits of some marker genes, visualized with <code class="docutils literal notranslate"><span class="pre">scv.pl.velocity(adata,</span> <span class="pre">gene_names)</span></code> or <code class="docutils literal notranslate"><span class="pre">scv.pl.scatter(adata,</span> <span class="pre">gene_names)</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">velocity</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;Cpe&#39;</span><span class="p">,</span>  <span class="s1">&#39;Gnao1&#39;</span><span class="p">,</span> <span class="s1">&#39;Ins2&#39;</span><span class="p">,</span> <span class="s1">&#39;Adk&#39;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_59_0.png" src="_images/03-velocity-review_59_0.png" />
</div>
</div>
<p>The black line corresponds to the estimated ‘steady-state’ ratio, i.e. the ratio of unspliced to spliced mRNA abundance which is in a constant transcriptional state. RNA velocity for a particular gene is determined as the residual, i.e. how much an observation deviates from that steady-state line. Positive velocity indicates that a gene is up-regulated, which occurs for cells that show higher abundance of unspliced mRNA for that gene than expected in steady state. Conversely, negative velocity
indicates that a gene is down-regulated.</p>
<p>For instance <em>Cpe</em> explains the directionality in the up-regulated Ngn3 (yellow) to Pre-endocrine (orange) to β-cells (green), while <em>Adk</em> explains the directionality in the down-regulated Ductal (dark green) to Ngn3 (yellow) to the remaining endocrine cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;Cpe&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">,</span> <span class="s1">&#39;velocity&#39;</span><span class="p">],</span>
               <span class="n">add_outline</span><span class="o">=</span><span class="s1">&#39;Ngn3 high EP, Pre-endocrine, Beta&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_61_0.png" src="_images/03-velocity-review_61_0.png" />
</div>
</div>
</section>
</section>
<section id="Postprocessing">
<h2>Postprocessing<a class="headerlink" href="#Postprocessing" title="Permalink to this headline">¶</a></h2>
<section id="Identify-important-genes">
<h3>Identify important genes<a class="headerlink" href="#Identify-important-genes" title="Permalink to this headline">¶</a></h3>
<p>We need a systematic way to identify genes that may help explain the resulting vector field and inferred lineages. To do so, we can test which genes have cluster-specific differential velocity expression, being siginificantly higher/lower compared to the remaining population. The module <code class="docutils literal notranslate"><span class="pre">scv.tl.rank_velocity_genes</span></code> runs a differential velocity t-test and outpus a gene ranking for each cluster. Thresholds can be set (e.g. <code class="docutils literal notranslate"><span class="pre">min_corr</span></code>) to restrict the test on a selection of gene candidates.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_velocity_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;clusters&#39;</span><span class="p">,</span> <span class="n">min_corr</span><span class="o">=.</span><span class="mi">3</span><span class="p">)</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;rank_velocity_genes&#39;</span><span class="p">][</span><span class="s1">&#39;names&#39;</span><span class="p">])</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ranking velocity genes
    finished (0:00:02) --&gt; added
    &#39;rank_velocity_genes&#39;, sorted scores by group ids (adata.uns)
    &#39;spearmans_score&#39;, spearmans correlation scores (adata.var)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ductal</th>
      <th>Ngn3 low EP</th>
      <th>Ngn3 high EP</th>
      <th>Pre-endocrine</th>
      <th>Beta</th>
      <th>Alpha</th>
      <th>Delta</th>
      <th>Epsilon</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Notch2</td>
      <td>Ptpn3</td>
      <td>Pde1c</td>
      <td>Pam</td>
      <td>Pax6</td>
      <td>Zcchc16</td>
      <td>Zdbf2</td>
      <td>Tmcc3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Sox5</td>
      <td>Hacd1</td>
      <td>Ptprs</td>
      <td>Sdk1</td>
      <td>Unc5c</td>
      <td>Nlgn1</td>
      <td>Spock3</td>
      <td>Heg1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Krt19</td>
      <td>Hspa8</td>
      <td>Pclo</td>
      <td>Baiap3</td>
      <td>Nnat</td>
      <td>Nell1</td>
      <td>Akr1c19</td>
      <td>Gpr179</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Hspa8</td>
      <td>Gm8113</td>
      <td>Rap1gap2</td>
      <td>Abcc8</td>
      <td>Tmem108</td>
      <td>Prune2</td>
      <td>Ptprt</td>
      <td>Ica1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Ano6</td>
      <td>Kcnq1</td>
      <td>Ttyh2</td>
      <td>Gnas</td>
      <td>Ptprt</td>
      <td>Ksr2</td>
      <td>Snap25</td>
      <td>Ncoa7</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span>
              <span class="n">add_outline</span><span class="o">=</span><span class="s1">&#39;Ngn3 high EP, Pre-endocrine, Beta&#39;</span><span class="p">)</span>

<span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Ngn3 high EP&#39;</span><span class="p">][:</span><span class="mi">5</span><span class="p">],</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Ngn3 high EP&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Pre-endocrine&#39;</span><span class="p">][:</span><span class="mi">5</span><span class="p">],</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Pre-endocrine&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_65_0.png" src="_images/03-velocity-review_65_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_65_1.png" src="_images/03-velocity-review_65_1.png" />
</div>
</div>
<p>The genes <em>Ptprs, Pclo, Pam, Abcc8, Gnas</em>, for instance, support the directionality from <strong>Ngn3 high EP</strong> (yellow) to <strong>Pre-endocrine</strong> (orange) to <strong>Beta</strong> (green).</p>
</section>
<section id="Velocities-in-cycling-progenitors">
<h3>Velocities in cycling progenitors<a class="headerlink" href="#Velocities-in-cycling-progenitors" title="Permalink to this headline">¶</a></h3>
<p>The cell cycle detected by RNA velocity, is biologically affirmed by cell cycle scores (standardized scores of mean expression levels of phase marker genes).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes_cell_cycle</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color_gradients</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;S_score&#39;</span><span class="p">,</span> <span class="s1">&#39;G2M_score&#39;</span><span class="p">],</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">perc</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">95</span><span class="p">])</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
calculating cell cycle phase
--&gt;     &#39;S_score&#39; and &#39;G2M_score&#39;, scores of cell cycle phases (adata.obs)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_69_1.png" src="_images/03-velocity-review_69_1.png" />
</div>
</div>
<p>For the cycling Ductal cells, we may screen through S and G2M phase markers. The previous module also computed a spearmans correlation score, which we can use to rank/sort the phase marker genes to then display their phase portraits.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">s_genes</span><span class="p">,</span> <span class="n">g2m_genes</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">get_phase_marker_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">s_genes</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">get_df</span><span class="p">(</span><span class="n">adata</span><span class="p">[:,</span> <span class="n">s_genes</span><span class="p">],</span> <span class="s1">&#39;spearmans_score&#39;</span><span class="p">,</span> <span class="n">sort_values</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>
<span class="n">g2m_genes</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">get_df</span><span class="p">(</span><span class="n">adata</span><span class="p">[:,</span> <span class="n">g2m_genes</span><span class="p">],</span> <span class="s1">&#39;spearmans_score&#39;</span><span class="p">,</span> <span class="n">sort_values</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>

<span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;cell cycle genes&#39;</span><span class="p">)</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">s_genes</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">g2m_genes</span><span class="p">[:</span><span class="mi">3</span><span class="p">]),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_71_0.png" src="_images/03-velocity-review_71_0.png" />
</div>
</div>
<p>Particularly <em>Hells</em> and <em>Top2a</em> are well-suited to explain the vector field in the cycling progenitors. <em>Top2a</em> gets assigned a high velocity shortly before it actually peaks in the G2M phase. There, the negative velocity then perfectly matches the immediately following down-regulation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">velocity</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;Hells&#39;</span><span class="p">,</span> <span class="s1">&#39;Top2a&#39;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_73_0.png" src="_images/03-velocity-review_73_0.png" />
</div>
</div>
</section>
<section id="Speed-and-coherence">
<h3>Speed and coherence<a class="headerlink" href="#Speed-and-coherence" title="Permalink to this headline">¶</a></h3>
<p>Two more useful stats: - The speed or rate of differentiation is given by the length of the velocity vector. - The coherence of the vector field (i.e., how a velocity vector correlates with its neighboring velocities) provides a measure of confidence.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">velocity_confidence</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">keys</span> <span class="o">=</span> <span class="s1">&#39;velocity_length&#39;</span><span class="p">,</span> <span class="s1">&#39;velocity_confidence&#39;</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">keys</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> <span class="n">perc</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">95</span><span class="p">])</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--&gt; added &#39;velocity_length&#39; (adata.obs)
--&gt; added &#39;velocity_confidence&#39; (adata.obs)
--&gt; added &#39;velocity_confidence_transition&#39; (adata.obs)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_75_1.png" src="_images/03-velocity-review_75_1.png" />
</div>
</div>
<p>These provide insights where cells differentiate at a slower/faster pace, and where the direction is un-/determined.</p>
<p>On cluster-level, we find that differentiation substantially speeds up after cell cycle exit (Ngn3 low EP), keeping the pace during Beta cell production while slowing down during Alpha cell production.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;clusters&#39;</span><span class="p">)[</span><span class="n">keys</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>
<span class="n">df</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style  type="text/css" >
    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col0 {
            background-color:  #3b4cc0;
            color:  #f1f1f1;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col1 {
            background-color:  #4c66d6;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col2 {
            background-color:  #d95847;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col3 {
            background-color:  #d44e41;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col4 {
            background-color:  #b40426;
            color:  #f1f1f1;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col5 {
            background-color:  #ebd3c6;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col6 {
            background-color:  #84a7fc;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row0_col7 {
            background-color:  #ebd3c6;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col0 {
            background-color:  #c3d5f4;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col1 {
            background-color:  #a5c3fe;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col2 {
            background-color:  #b40426;
            color:  #f1f1f1;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col3 {
            background-color:  #f7b497;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col4 {
            background-color:  #7093f3;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col5 {
            background-color:  #9ebeff;
            color:  #000000;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col6 {
            background-color:  #3b4cc0;
            color:  #f1f1f1;
        }    #T_a364a438_962f_11ea_8edb_8c859086e093row1_col7 {
            background-color:  #d1dae9;
            color:  #000000;
        }</style><table id="T_a364a438_962f_11ea_8edb_8c859086e093" ><thead>    <tr>        <th class="index_name level0" >clusters</th>        <th class="col_heading level0 col0" >Ductal</th>        <th class="col_heading level0 col1" >Ngn3 low EP</th>        <th class="col_heading level0 col2" >Ngn3 high EP</th>        <th class="col_heading level0 col3" >Pre-endocrine</th>        <th class="col_heading level0 col4" >Beta</th>        <th class="col_heading level0 col5" >Alpha</th>        <th class="col_heading level0 col6" >Delta</th>        <th class="col_heading level0 col7" >Epsilon</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_a364a438_962f_11ea_8edb_8c859086e093level0_row0" class="row_heading level0 row0" >velocity_length</th>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col0" class="data row0 col0" >5.707904</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col1" class="data row0 col1" >6.227023</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col2" class="data row0 col2" >13.257741</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col3" class="data row0 col3" >13.429206</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col4" class="data row0 col4" >14.219882</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col5" class="data row0 col5" >10.491913</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col6" class="data row0 col6" >7.623143</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row0_col7" class="data row0 col7" >10.477606</td>
            </tr>
            <tr>
                        <th id="T_a364a438_962f_11ea_8edb_8c859086e093level0_row1" class="row_heading level0 row1" >velocity_confidence</th>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col0" class="data row1 col0" >0.775042</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col1" class="data row1 col1" >0.756474</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col2" class="data row1 col2" >0.896944</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col3" class="data row1 col3" >0.830438</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col4" class="data row1 col4" >0.725973</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col5" class="data row1 col5" >0.752715</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col6" class="data row1 col6" >0.691188</td>
                        <td id="T_a364a438_962f_11ea_8edb_8c859086e093row1_col7" class="data row1 col7" >0.784467</td>
            </tr>
    </tbody></table></div>
</div>
</section>
<section id="Velocity-graph-and-pseudotime">
<h3>Velocity graph and pseudotime<a class="headerlink" href="#Velocity-graph-and-pseudotime" title="Permalink to this headline">¶</a></h3>
<p>We can visualize the velocity graph to portray all velocity-inferred cell-to-cell connections/transitions. It can be confined to high-probability transitions by setting a <code class="docutils literal notranslate"><span class="pre">threshold</span></code>. The graph, for instance, indicates two phases of Epsilon cell production, coming from early and late Pre-endocrine cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">velocity_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=.</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_79_0.png" src="_images/03-velocity-review_79_0.png" />
</div>
</div>
<p>Further, the graph can be used to draw descendents/anscestors coming from a specified cell. Here, a pre-endocrine cell is traced to its potential fate.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">get_cell_transitions</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">starting_cell</span><span class="o">=</span><span class="mi">70</span><span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">velocity_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;lightgrey&#39;</span><span class="p">,</span> <span class="n">edge_width</span><span class="o">=.</span><span class="mi">05</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">120</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;ascending&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;gnuplot&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_81_0.png" src="_images/03-velocity-review_81_0.png" />
</div>
</div>
<p>Finally, based on the velocity graph, a <em>velocity pseudotime</em> can be computed. After inferring a distribution over root cells from the graph, it measures the average number of steps it takes to reach a cell after walking along the graph starting from the root cells.</p>
<p>Contrarily to diffusion pseudotime, it implicitly infers the root cells and is based on the directed velocity graph instead of the similarity-based diffusion kernel.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">velocity_pseudotime</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;velocity_pseudotime&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;gnuplot&#39;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
computing terminal states
    identified 2 regions of root cells and 1 region of end points
    finished (0:00:00) --&gt; added
    &#39;root_cells&#39;, root cells of Markov diffusion process (adata.obs)
    &#39;end_points&#39;, end points of Markov diffusion process (adata.obs)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_83_1.png" src="_images/03-velocity-review_83_1.png" />
</div>
</div>
</section>
<section id="PAGA-velocity-graph">
<h3>PAGA velocity graph<a class="headerlink" href="#PAGA-velocity-graph" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://doi.org/10.1186/s13059-019-1663-x">PAGA</a> graph abstraction has benchmarked as top-performing method for trajectory inference. It provides a graph-like map of the data topology with weighted edges corresponding to the connectivity between two clusters. Here, PAGA is extended by velocity-inferred directionality.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># PAGA requires to install igraph, if not done yet.</span>
<span class="o">!</span>pip install python-igraph --upgrade --quiet
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># this is needed due to a current bug - bugfix is coming soon.</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;neighbors&#39;</span><span class="p">][</span><span class="s1">&#39;distances&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;distances&#39;</span><span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;neighbors&#39;</span><span class="p">][</span><span class="s1">&#39;connectivities&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;connectivities&#39;</span><span class="p">]</span>

<span class="n">scv</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="s1">&#39;clusters&#39;</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">get_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;paga/transitions_confidence&#39;</span><span class="p">,</span> <span class="n">precision</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="n">df</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{:.2g}</span><span class="s1">&#39;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
running PAGA
    finished (0:00:01) --&gt; added
    &#39;paga/transitions_confidence&#39;, connectivities adjacency (adata.uns)
    &#39;paga/connectivities&#39;, connectivities adjacency (adata.uns)
    &#39;paga/connectivities_tree&#39;, connectivities subtree (adata.uns)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style  type="text/css" >
    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col1 {
            background-color:  #08306b;
            color:  #f1f1f1;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col2 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col3 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col4 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col5 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col6 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col7 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col1 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col2 {
            background-color:  #08306b;
            color:  #f1f1f1;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col3 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col4 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col5 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col6 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col7 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col1 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col2 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col3 {
            background-color:  #08306b;
            color:  #f1f1f1;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col4 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col5 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col6 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col7 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col1 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col2 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col3 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col4 {
            background-color:  #08306b;
            color:  #f1f1f1;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col5 {
            background-color:  #08306b;
            color:  #f1f1f1;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col6 {
            background-color:  #08306b;
            color:  #f1f1f1;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col7 {
            background-color:  #08306b;
            color:  #f1f1f1;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col1 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col2 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col3 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col4 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col5 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col6 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col7 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col1 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col2 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col3 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col4 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col5 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col6 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col7 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col1 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col2 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col3 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col4 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col5 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col6 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col7 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col0 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col1 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col2 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col3 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col4 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col5 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col6 {
            background-color:  #f7fbff;
            color:  #000000;
        }    #T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col7 {
            background-color:  #f7fbff;
            color:  #000000;
        }</style><table id="T_a6afbe02_962f_11ea_8edb_8c859086e093" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >Ductal</th>        <th class="col_heading level0 col1" >Ngn3 low EP</th>        <th class="col_heading level0 col2" >Ngn3 high EP</th>        <th class="col_heading level0 col3" >Pre-endocrine</th>        <th class="col_heading level0 col4" >Beta</th>        <th class="col_heading level0 col5" >Alpha</th>        <th class="col_heading level0 col6" >Delta</th>        <th class="col_heading level0 col7" >Epsilon</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row0" class="row_heading level0 row0" >Ductal</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col0" class="data row0 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col1" class="data row0 col1" >0.15</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col2" class="data row0 col2" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col3" class="data row0 col3" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col4" class="data row0 col4" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col5" class="data row0 col5" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col6" class="data row0 col6" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row0_col7" class="data row0 col7" >0</td>
            </tr>
            <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row1" class="row_heading level0 row1" >Ngn3 low EP</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col0" class="data row1 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col1" class="data row1 col1" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col2" class="data row1 col2" >0.24</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col3" class="data row1 col3" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col4" class="data row1 col4" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col5" class="data row1 col5" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col6" class="data row1 col6" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row1_col7" class="data row1 col7" >0</td>
            </tr>
            <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row2" class="row_heading level0 row2" >Ngn3 high EP</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col0" class="data row2 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col1" class="data row2 col1" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col2" class="data row2 col2" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col3" class="data row2 col3" >0.22</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col4" class="data row2 col4" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col5" class="data row2 col5" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col6" class="data row2 col6" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row2_col7" class="data row2 col7" >0</td>
            </tr>
            <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row3" class="row_heading level0 row3" >Pre-endocrine</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col0" class="data row3 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col1" class="data row3 col1" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col2" class="data row3 col2" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col3" class="data row3 col3" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col4" class="data row3 col4" >0.48</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col5" class="data row3 col5" >0.12</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col6" class="data row3 col6" >0.21</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row3_col7" class="data row3 col7" >0.11</td>
            </tr>
            <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row4" class="row_heading level0 row4" >Beta</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col0" class="data row4 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col1" class="data row4 col1" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col2" class="data row4 col2" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col3" class="data row4 col3" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col4" class="data row4 col4" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col5" class="data row4 col5" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col6" class="data row4 col6" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row4_col7" class="data row4 col7" >0</td>
            </tr>
            <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row5" class="row_heading level0 row5" >Alpha</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col0" class="data row5 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col1" class="data row5 col1" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col2" class="data row5 col2" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col3" class="data row5 col3" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col4" class="data row5 col4" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col5" class="data row5 col5" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col6" class="data row5 col6" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row5_col7" class="data row5 col7" >0</td>
            </tr>
            <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row6" class="row_heading level0 row6" >Delta</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col0" class="data row6 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col1" class="data row6 col1" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col2" class="data row6 col2" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col3" class="data row6 col3" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col4" class="data row6 col4" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col5" class="data row6 col5" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col6" class="data row6 col6" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row6_col7" class="data row6 col7" >0</td>
            </tr>
            <tr>
                        <th id="T_a6afbe02_962f_11ea_8edb_8c859086e093level0_row7" class="row_heading level0 row7" >Epsilon</th>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col0" class="data row7 col0" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col1" class="data row7 col1" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col2" class="data row7 col2" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col3" class="data row7 col3" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col4" class="data row7 col4" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col5" class="data row7 col5" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col6" class="data row7 col6" >0</td>
                        <td id="T_a6afbe02_962f_11ea_8edb_8c859086e093row7_col7" class="data row7 col7" >0</td>
            </tr>
    </tbody></table></div>
</div>
<p>This reads from left/row to right/column, thus e.g. assigning a confident transition from Ductal to Ngn3 low EP.</p>
<p>This table can be summarized by a directed graph superimposed onto the UMAP embedding.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scv</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=.</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_edge_width</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">node_size_scale</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/03-velocity-review_88_0.png" src="_images/03-velocity-review_88_0.png" />
</div>
</div>
<script type="application/vnd.jupyter.widget-state+json">
{"state": {}, "version_major": 2, "version_minor": 0}
</script></section>
</section>
</section>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">

          <div class="sidebar-resize-handle"></div>

<h3><a href="index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="plotting.html">Plotting setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="01-anndata-review.html">AnnData review</a></li>
<li class="toctree-l1"><a class="reference internal" href="02-analysis-review.html">Basic single-cell analysis</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">RNA Velocity review</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Setup">Setup</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-libraries">Import libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Setup-plotting">Setup plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Utility-functions">Utility functions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Load-data">Load data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Load-AnnData-object">Load AnnData object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Detailed-view-of-adata">Detailed view of adata</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Show-spliced/unspliced">Show spliced/unspliced</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Preprocess-the-Data">Preprocess the Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Compute-RNA-velocity">Compute RNA velocity</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Estimate-RNA-velocity">Estimate RNA velocity</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Project-the-velocities">Project the velocities</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Interpret-the-velocities">Interpret the velocities</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Postprocessing">Postprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Identify-important-genes">Identify important genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Velocities-in-cycling-progenitors">Velocities in cycling progenitors</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Speed-and-coherence">Speed and coherence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Velocity-graph-and-pseudotime">Velocity graph and pseudotime</a></li>
<li class="toctree-l3"><a class="reference internal" href="#PAGA-velocity-graph">PAGA velocity graph</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

<hr class="docutils" />
<ul>
  <li class="toctree-l1"><a class="reference internal" href="genindex.html" accesskey="I">General Index</a></li>
</ul>
<div id="ethical-ad-placement"></div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>

    <nav class="relbar">
      <a class="previous" href="02-analysis-review.html" aria-label="Previous topic" >
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"/></svg>
        </div>
        <div class="title">
          <span class="text">
            <span class="direction">previous</span>
            Basic single-cell analysis
          </span>
        </div>
      </a>
      <a class="next" href="references.html" aria-label="Next topic" >
        <div class="title">
          <span class="text">
            <span class="direction">next</span>
            References
          </span>
        </div>
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"/></svg>
        </div>
      </a>
    </nav>

    <footer role="contentinfo">
      &#169; Copyright 2022, Cameron Smith.
      Created using <a class="reference external" href="https://www.sphinx-doc.org/">Sphinx</a> 2.4.4.
      <a class="reference external" href="https://insipid-sphinx-theme.readthedocs.io/">Insipid Theme</a>.
    </footer>
  </body>
</html>